---
title: "compare_simulated_data"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, results='hide', message=FALSE, warning=FALSE}
library(atgnat)
library(SingleCellExperiment)
library(tradeSeq)
library(slingshot)
library(scran)
library(scater)
library(BiocParallel)
library(dyngen)
library(bluster)
library(fs)
library(utils)
```

```{r randomseed}
RNGversion("3.5.0")
SEED = 7
set.seed(SEED)
```

In this article, we'll simulate a dataset, and show that BLASE works well with it, 
mapping pseudobulked SC counts to the same SC data. We split the simulated data in 
half and use one half of it as the scRNA-seq, and the other for pseudobulking.

```{r concurrency}
N_CORES = 4
if (ami::using_ci()) {
  N_CORES = 2
}
```

## Generate the simulation data
Based on the [dyngen Getting Started tutorial](https://dyngen.dynverse.org/articles/getting_started.html#step-1-define-backbone-and-other-parameters). 
```{r generateCounts}
root_dir = tools::R_user_dir("BLASE", "data")
article_dir = path(root_dir, "compare_simulated_data")

# TODO create an article with steps to create the simulated data and then just save as package data like we do for the MCA data
if (!dir.exists(article_dir)) {
  dir.create(article_dir, recursive = TRUE)
}

simulation_rds_path = path(article_dir, "simulation", ext="rds")
if (!file.exists(simulation_rds_path)) {
  # realnet for featurenetwork from realnets
  # realcount for generate experiment from realcounts
  
  simulations_num = 150
  backbone <- backbone_linear_simple()
  config <-
    initialise_model(
      backbone = backbone,
      num_cells = 2000,
      num_tfs = nrow(backbone$module_info),
      num_targets = 400,
      num_hks = 50,
      verbose = FALSE,
      feature_network_params = feature_network_default(
        realnet="regulatorycircuits_04_mesenchymal_mixed"
      ),
      experiment_params = experiment_snapshot(
        realcount="zenodo_1443566_real_silver_trophoblast-stem-cell-trophoblast-differentiation_mca"
      ),
      download_cache_dir = tools::R_user_dir("dyngen", "data"),
      simulation_params = simulation_default(
        total_time = 1000,
        ssa_algorithm = ssa_etl(tau = 30/3600),
        census_interval = 4,
        experiment_params = simulation_type_wild_type(num_simulations = simulations_num, seed=seq_len(simulations_num))
      ),
      num_cores=N_CORES
    )
  
  # Should be > 10 ideally
  floor(config$simulation_params$total_time / config$simulation_params$census_interval) * simulations_num / config$numbers$num_cells
  config$experiment_params$realcount
  config$feature_network_params$realnet
  
  model <- generate_tf_network(config)
  model <- generate_feature_network(model)
  model <- generate_kinetics(model)
  model <- generate_gold_standard(model)
  model <- generate_cells(model)
  model <- generate_experiment(model)
  
  saveRDS(model, simulation_rds_path);
} else {
    print("Using cached")
    model = readRDS(simulation_rds_path)
}

```

## Cluster SCE

Here we'll set up the SCE and preprocess the data ready for trajectory analysis. 
We show the `sim_time` mapped onto it, which shows the true simulated trajectory.

```{r clusterSCE}
sce <- as_sce(model)
rm(model)
gc()

sce <- computeSumFactors(sce)
sce <- logNormCounts(sce)
normcounts(sce) <- exp(logcounts(sce))
sce = runPCA(sce)

colLabels(sce) <- clusterCells(sce, use.dimred='PCA',
    BLUSPARAM=NNGraphParam(cluster.fun="louvain")) 

sce = runUMAP(sce)
gridExtra::grid.arrange(
  plotPCA(sce, colour_by="label"),
  plotUMAP(sce, colour_by="label"),
  plotPCA(sce, colour_by="sim_time"),
  plotUMAP(sce, colour_by="sim_time"),
  ncol=2
)

```
```{r}
START_CLUSTER = 4
```

## Split Datasets

We will now split the dataset so that the test dataset is unseen.

```{r}
sce$useAsBulk = round(runif(n=ncol(sce), min=0, max=1)) == 1
bulk_sce = subset(sce, , useAsBulk==TRUE)
sc_sce = subset(sce, , useAsBulk==FALSE)
rm(sce)
```

## SC Trajectory Analysis
Now we'll use slingshot to infer a trajectory, and tradeseq to identify highly variable genes. We will compare the trajectory to the simulated time below.
```{r}
sc_sce

sc_sce <- slingshot(sc_sce, reducedDim = 'UMAP', clusterLabels='label', start.clus=START_CLUSTER)

gridExtra::grid.arrange(
  plotPCA(sc_sce, colour_by="sim_time"),
  plotPCA(sc_sce, colour_by="slingPseudotime_1"),
  plotUMAP(sc_sce, colour_by="sim_time"),
  plotUMAP(sc_sce, colour_by="slingPseudotime_1"),
  plotPCA(sc_sce, colour_by="label"),
  plotUMAP(sc_sce, colour_by="label"),
  ncol=2
)

sc_sce <- fitGAM(sc_sce, parallel=T, BPPARAM=MulticoreParam(N_CORES), nknots=4)
associationTestResult <- associationTest(sc_sce, lineages=T, global=F, contrastType="consecutive")

```

## Analyse with BLASE

We start by tuning the parameters for BLASE. See the [vignette](articles/assign-bulk-to-pseudotime.html) for more details.

```{r}
genelist = atgnat::get_top_n_genes(associationTestResult, n_genes = 200, lineage = 1)
res = find_best_params(sc_sce,
                       genelist,
                       split_by="pseudotime_range",
                       pseudotime_slot="slingPseudotime_1",
                       bins_count_range = seq.int(3, 10, 1),
                       gene_count_range = seq.int(3, 60, 3))
plot_find_best_params_results(res)
```
```{r}
genes = 30
bins =5
atgnatData = as.AtgnatData(sc_sce, pseudotime_slot="slingPseudotime_1", n_bins=bins, split_by="pseudotime_range")
atgnatData@genes = genelist[1:genes]
evaluate_parameters(atgnatData, make_plot = TRUE)
```
## Prepare Bulk

Here we set up the same trajectory with the bulk dataset, and split it into separate "bulk" counts.

```{r}
bulk_sce

bulk_sce <- slingshot(bulk_sce, reducedDim = 'UMAP', clusterLabels='label', start.clus=START_CLUSTER)

gridExtra::grid.arrange(
  plotPCA(bulk_sce, colour_by="sim_time"),
  plotPCA(bulk_sce, colour_by="slingPseudotime_1"),
  plotUMAP(bulk_sce, colour_by="sim_time"),
  plotUMAP(bulk_sce, colour_by="slingPseudotime_1"),
  ncol=2
)

bulk_sce = assign_pseudotime_bins(bulk_sce, split_by='pseudotime_range', n_bins = bins, pseudotime_slot="slingPseudotime_1")
bulk_sce$replicate = round(runif(n=ncol(bulk_sce), min=1, max=2))
bulkified_bins = atgnat::get_bins_as_bulk(bulk_sce, min_cells_for_bulk = 20)
```

## Map the bulk to the single cell

Now we'll map the bulk onto the single cell data with BLASE. We can see the bins and mapping results below. Most are confidently mapped correctly, and 4 are not confidently mapped, but still correctly.

```{r}
mapping_results = c()
for (bulkname in colnames(bulkified_bins)) {
  res = map_best_bin(atgnatData, bulkname, bulkified_bins)
  mapping_results = c(mapping_results, res)
  print(res)
}

sc_sce = assign_pseudotime_bins(sc_sce, split_by = 'pseudotime_range', n_bins = bins, pseudotime_slot='slingPseudotime_1')
gridExtra::grid.arrange(
  plotUMAP(bulk_sce, colour_by="sim_time"),
  plotUMAP(bulk_sce, colour_by="slingPseudotime_1"),
  plotUMAP(sc_sce, colour="pseudotime_bin"),
  ncol=3
)
atgnat::plot_mapping_result_heatmap(rev(mapping_results))
```
Now we can look at the population proportions of the bins:
```{r}
gridExtra::grid.arrange(
  plotUMAP(sc_sce, colour_by="label"),
  plotUMAP(sc_sce, colour="pseudotime_bin"),
  plotUMAP(bulk_sce, colour_by="label"),
  plotUMAP(bulk_sce, colour="pseudotime_bin"),
  ncol=2
)

gridExtra::grid.arrange(
  atgnat::plot_bin_population(sc_sce, 1, group_by_slot="label"),
  atgnat::plot_bin_population(bulk_sce, 1, group_by_slot="label"),
  atgnat::plot_mapping_result_corr(mapping_results[[1]]),
  atgnat::plot_mapping_result_corr(mapping_results[[2]]),
  atgnat::plot_bin_population(sc_sce, 2, group_by_slot="label"),
  atgnat::plot_bin_population(bulk_sce, 2, group_by_slot="label"),
  atgnat::plot_mapping_result_corr(mapping_results[[3]]),
  atgnat::plot_mapping_result_corr(mapping_results[[4]]),
  atgnat::plot_bin_population(sc_sce, 3, group_by_slot="label"),
  atgnat::plot_bin_population(bulk_sce, 3, group_by_slot="label"),
  atgnat::plot_mapping_result_corr(mapping_results[[5]]),
  atgnat::plot_mapping_result_corr(mapping_results[[6]]),
  atgnat::plot_bin_population(sc_sce, 4, group_by_slot="label"),
  atgnat::plot_bin_population(bulk_sce, 4, group_by_slot="label"),
  atgnat::plot_mapping_result_corr(mapping_results[[7]]),
  atgnat::plot_mapping_result_corr(mapping_results[[8]]),
  atgnat::plot_bin_population(sc_sce, 5, group_by_slot="label"),
  atgnat::plot_bin_population(bulk_sce, 5, group_by_slot="label"),
  atgnat::plot_mapping_result_corr(mapping_results[[9]]),
  atgnat::plot_mapping_result_corr(mapping_results[[10]]),
  ncol=4
)

```
## Session Info
```{r sessionInfo}
sessionInfo()
```
