---
title: "DE With BLASE"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, results='hide', message=FALSE, warning=FALSE}
library(readxl)
library(scran)
library(slingshot)
library(tradeSeq)
library(BiocParallel)
library(fs)
library(gridExtra)
library(scater)
library(utils)
library(blase)
library(ami)
library(plyr)
library(scales)
library(reshape2)
```

```{r randomseed}
RNGversion("3.5.0")
SEED = 7
set.seed(SEED)
```

```{r concurrency}
N_CORES = 4
if (ami::using_ci()) {
  N_CORES = 2
}
```

In this vignette we will use BLASE to map bulk Plasmodium Falciparum data from an experiment ([Zhang et al. 2021](https://doi.org/10.1038/s41467-021-24814-1)) on heat shock response in P. Falciparum, onto the Malaria Cell Atlas ([Howick et al. 2019](https://doi.org/10.1126/science.aaw2619), 7G8 & NF54 strains). We look at Zhang et al's wild type control data (from the NF54 strain) as well as the knockouts (pb4 and pb31) in this article. The paper reported that the knockouts both handled the heat shock less well than the wild type strain, and that pb31 was most susceptible to heat shock, which we will see in our analysis below.

The controls kept all strains at 37 degrees, whereas the heat shock was applied to each strain as 3 8-hour increases to 41 degrees, simulating malarial fever.

## Fetch the data
```{r getData}
root_dir = tools::R_user_dir("BLASE", "data")
article_dir = path(root_dir, "DE_with_blase")

if (!dir.exists(article_dir)) {
  dir.create(article_dir, recursive = TRUE)
}

bulk_path = path(article_dir, "heat_shock_bulks", ext="xlsx")
if (!file.exists(bulk_path)) {
  download.file('https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-021-24814-1/MediaObjects/41467_2021_24814_MOESM5_ESM.xlsx', bulk_path)
}else {
  print("Using cached")
}

# FPKM normalised
bulk = as.data.frame(read_excel(bulk_path, range = 'S3A_RNAseq!A4:P2571'))
rownames(bulk) = bulk$`Transcript_ID`
rownames(bulk) = gsub(x=rownames(bulk), pattern="_", replacement="-", fixed=TRUE)

bulk = subset(bulk, , -c(`GeneID`, `Transcript_ID`, `Gene.NameSymbol`, `TranscriptProductDescription`))

genes_to_fix_bulk = rownames(bulk)[!(rownames(bulk) %in% sub(x=rownames(bulk), pattern="\\.[0-9]", replacement=""))]
genes_to_fix_bulk = unique(sub(x=genes_to_fix_bulk, pattern="\\.[0-9]", replacement=""))
```

```{R removeBulkDupTranscripts}
new_rows = data.frame()
rownames_to_remove = c()

# Do some pre-work to get colSums quickly
bulk_matrix = data.matrix(bulk)
bulk_n = ncol(bulk_matrix)

for (gene in genes_to_fix_bulk) {
  targetRowNames = rownames(bulk)[grep(x=rownames(bulk), pattern=gene)]
  rownames_to_remove = c(rownames_to_remove, targetRowNames)
  # N = col index, M = row index
  counts = .colSums(bulk_matrix[targetRowNames,], m=length(targetRowNames), n=bulk_n)
  counts = t(data.frame(counts=counts))
  rownames(counts) = c(gene)
  colnames(counts) = colnames(bulk)
  new_rows = rbind(new_rows, counts)
}

bulk = bulk[ !(rownames(bulk) %in% rownames_to_remove), ]
bulk = rbind(bulk, new_rows)

genes_to_fix_bulk = rownames(bulk)[!(rownames(bulk) %in% sub(x=rownames(bulk), pattern="\\.[0-9]", replacement=""))]
genes_to_fix_bulk = unique(sub(x=genes_to_fix_bulk, pattern="\\.[0-9]", replacement=""))
print(paste("Remaining genes with dups:", genes_to_fix_bulk))

rm (new_rows, rownames_to_remove, bulk_matrix, bulk_n, genes_to_fix_bulk, counts)
gc()
```
## Prepare SC

We are using the Malaria Cell Atlas data which we have pre-processed (here)[articles/generate-MCA-pf-object.html]. This data is available for your to experiment with as part of the package.

```{r loadSC}
data(processed_MCA_PF_SCE, package="blase")
gridExtra::grid.arrange(
  plotPCA(processed_MCA_PF_SCE, colour_by="STAGE_LR"),
  plotUMAP(processed_MCA_PF_SCE, colour_by="STAGE_LR"),
  plotPCA(processed_MCA_PF_SCE, colour_by="DAY"),
  plotUMAP(processed_MCA_PF_SCE, colour_by="DAY"),
  plotPCA(processed_MCA_PF_SCE, colour_by="STRAIN"),
  plotUMAP(processed_MCA_PF_SCE, colour_by="STRAIN"),
  plotPCA(processed_MCA_PF_SCE, colour_by="slingPseudotime_1"),
  plotUMAP(processed_MCA_PF_SCE, colour_by="slingPseudotime_1"),
  ncol=2
)
```

```{r tradeSeq}
associationTestResult <- associationTest(processed_MCA_PF_SCE, lineages=T, global=F, contrastType="consecutive")
```

## Prepare BLASE
We prepare BLASE by checking for a good number of genes and bins.
```{r}
genelist = blase::get_top_n_genes(associationTestResult, n_genes = 500, lineage = 1)

res = find_best_params(processed_MCA_PF_SCE,
                       genelist,
                       split_by="pseudotime_range",
                       pseudotime_slot="slingPseudotime_1",
                       bins_count_range = c(6, 8, 10),
                       gene_count_range = seq.int(250, 340, 3))
plot_find_best_params_results(res)
```

```{r}
#cat(gsub(x=paste0("", genelist), pattern="-", replacement = "_"), sep = "\n")

#waves = blase::get_waves(processed_MCA_PF_SCE)

#waves[c("PF3D7-0831700","PF3D7-1116800"),]

#blase::gene_selection_matrix(processed_MCA_PF_SCE, genes=c(
#  "PF3D7-0831700", 
#  "PF3D7-1116800",
#  "PF3D7-1014100",
#  "PF3D7-0303200",
#  "PF3D7-0701900"
#  ), waves=waves, target_matrix_size=5)

#length(genelist)
#genelist = genelist[!(genelist %in% c("PF3D7-0831700","PF3D7-1116800"))]
#length(genelist)
```

```{r}
genes = 300
bins = 8
blaseData = as.BlaseData(processed_MCA_PF_SCE, pseudotime_slot="slingPseudotime_1", n_bins=bins, split_by="pseudotime_range")
blaseData@genes = genelist[1:genes]
evaluate_parameters(blaseData, make_plot = TRUE)
processed_MCA_PF_SCE = assign_pseudotime_bins(processed_MCA_PF_SCE, pseudotime_slot="slingPseudotime_1", n_bins=bins, split_by="pseudotime_range")
plotUMAP(processed_MCA_PF_SCE, color="pseudotime_bin")
gridExtra::grid.arrange(
  plotUMAP(processed_MCA_PF_SCE, color="STAGE_LR"),
  plotUMAP(processed_MCA_PF_SCE, color="pseudotime_bin"),
  ncol=2
)
```
```{r}
#waves = get_waves(processed_MCA_PF_SCE, n_cores = N_CORES)
#fourier_genes = blase::select_genes_by_fourier_method(processed_MCA_PF_SCE, waves, n_groups=100, top_n_per_group=5)
#fourier_genes
#gene_selection_matrix(processed_MCA_PF_SCE, waves, genes=rownames(fourier_genes), target_matrix_size=90)

#dim(fourier_genes)
#fourier_genes = fourier_genes[ !(rownames(fourier_genes) %in%  c("PF3D7-0831700","PF3D7-1116800", "PF3D7_0201800")), ]
#dim(fourier_genes)

#blaseData@genes = rownames(fourier_genes)
#cat(gsub(x=paste0("", rownames(fourier_genes)), pattern="-", replacement = "_"), sep = "\n")

```

And we can see these genes arranged over pseudotime:

```{r}
waves = get_waves(processed_MCA_PF_SCE)
gene_selection_matrix(processed_MCA_PF_SCE, waves, genes=blaseData@genes, target_matrix_size=length(unique(blaseData@genes)))
```

## Map Bulk data onto Single Cell
Now we can map the bulk data onto the single cell.
```{r}
mapping_results = c()
for (bulkname in colnames(bulk)) {
  res = map_best_bin(blaseData, bulkname, bulk)
  mapping_results = c(mapping_results, res)
}

blase::plot_mapping_result_heatmap(rev(mapping_results), annotate = TRUE)
```

We can see the wild type develops more normally than the 2 knockouts, who show delayed growth generally, and less of a response to the heat shock (41 degrees).

```{r}
# Wildtype
gridExtra::grid.arrange(
  blase::plot_mapping_result_heatmap(rev(mapping_results[1:2]), annotate = TRUE), # 1
  blase::plot_mapping_result_heatmap(rev(mapping_results[3:4]), annotate = TRUE), # 2
  blase::plot_bin_population(processed_MCA_PF_SCE, 2, group_by_slot="STAGE_LR"), # 1
  blase::plot_bin_population(processed_MCA_PF_SCE, 4, group_by_slot="STAGE_LR"), # 2
  blase::plot_bin_population(processed_MCA_PF_SCE, 3, group_by_slot="STAGE_LR"), # 1
  blase::plot_bin_population(processed_MCA_PF_SCE, 4, group_by_slot="STAGE_LR"), # 2
  ncol=2
)

# pb4 knockout
gridExtra::grid.arrange(
  blase::plot_mapping_result_heatmap(rev(mapping_results[5:6]), annotate = TRUE), # 1
  blase::plot_mapping_result_heatmap(rev(mapping_results[7:8]), annotate = TRUE), # 2
  blase::plot_bin_population(processed_MCA_PF_SCE, 2, group_by_slot="STAGE_LR"), # 1
  blase::plot_bin_population(processed_MCA_PF_SCE, 3, group_by_slot="STAGE_LR"), # 2
  ggplot() + theme_void(), # 1
  ggplot() + theme_void(), # 2
  ncol=2
)

# pb31 knockout
gridExtra::grid.arrange(
  blase::plot_mapping_result_heatmap(rev(mapping_results[9:10]), annotate = TRUE), # 1
  blase::plot_mapping_result_heatmap(rev(mapping_results[11:12]), annotate = TRUE), # 2
  blase::plot_bin_population(processed_MCA_PF_SCE, 1, group_by_slot="STAGE_LR"), # 1
  blase::plot_bin_population(processed_MCA_PF_SCE, 3, group_by_slot="STAGE_LR"), # 2
  blase::plot_bin_population(processed_MCA_PF_SCE, 2, group_by_slot="STAGE_LR"), # 1
  ggplot() + theme_void(), #2
  ncol=2
)
```

## Cibersort Analysis for comparison

Cibersortx is another tool which could be used for this kind of analysis. Let's see how it performs compared to BLASE. Below is code for producing the files we used to analyse with cibersortx.

```{r, eval=FALSE}
cibersort_bulk = bulk
rownames(cibersort_bulk) = gsub(x=rownames(cibersort_bulk), pattern="-", replacement = ".")
colnames(cibersort_bulk) = gsub(x=colnames(cibersort_bulk), pattern="_", replacement = ".")
cibersort_bulk = cbind(genes = rownames(cibersort_bulk), cibersort_bulk, row.names = NULL)

cibersort_sc = normcounts(processed_MCA_PF_SCE)
rownames(cibersort_sc) = gsub(x=rownames(cibersort_sc), pattern="-", replacement = ".")
colnames(cibersort_sc) = processed_MCA_PF_SCE@colData$pseudotime_bin
cibersort_sc = cbind(genes = rownames(cibersort_sc), cibersort_sc, row.names = NULL)

all(colnames(normcounts(processed_MCA_PF_SCE)) == rownames(processed_MCA_PF_SCE@colData))

#write.table(cibersort_bulk, file = "cibersort_bulk_de.tsv", sep = "\t", row.names=FALSE, col.names=TRUE)
#write.table(cibersort_sc, file = "cibersort_sc_de.tsv", sep = "\t", row.names=FALSE, col.names=TRUE)
```
After analysing with cibersortx, using default settings, and 100 iterations for confidence, the results are given in this file.
```{r}
data_dir <- system.file("extdata/", package = "blase")
path_to_cibersort_results = path(data_dir, "CIBERSORTX_Job6_Results", ext = "csv")

cibersort_results = read.table(path_to_cibersort_results, sep = ',', header=TRUE)
rownames(cibersort_results) = cibersort_results$Mixture
cibersort_results <- subset( cibersort_results, select = -Mixture )

cibersort_results_calls = cibersort_results[c('X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8')]
cibersort_results_quality = cibersort_results[c('P.value', 'Correlation', 'RMSE')]

to_plot = cibersort_results_calls
to_plot$BulkID <- rownames(cibersort_results_calls)
to_plot <- melt(to_plot)
p1=ggplot(to_plot, aes(variable, BulkID)) +
  geom_tile(aes(fill = value), colour = "white") +
  geom_text(aes(label = round(value, 2))) +
  scale_fill_gradient(low = "white", high = "red")

cibersort_results_quality

to_plot = cibersort_results_quality
to_plot$BulkID <- rownames(cibersort_results_quality)
to_plot <- melt(to_plot)
to_plot <- ddply(to_plot, .(variable), transform, rescale = rescale(value))
p2=ggplot(to_plot, aes(variable, BulkID)) +
  geom_tile(aes(fill = rescale), colour = "white") +
  geom_text(aes(label = round(value, 3))) +
  scale_fill_gradient(low = "white", high = "red")

p1
p2
```




## Session Info
```{r sessionInfo}
sessionInfo()
```
