<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Assigning bulk RNA-seq to pseudotime • blase</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assigning bulk RNA-seq to pseudotime">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/How-Blase-Works.html">How BLASE Works</a></li>
    <li><a class="dropdown-item" href="../articles/assign-bulk-to-pseudotime.html">Assigning Bulk to Pseudotime</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Articles</h6></li>
    <li><a class="dropdown-item" href="../articles/BLASE-for-annotating-scRNA-seq.html">BLASE for annotating scRNA-seq</a></li>
    <li><a class="dropdown-item" href="../articles/BLASE-for-excluding-developmental-genes-from-bulk-RNA-seq.html">BLASE for excluding developmental genes from bulk RNA-seq</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Assigning bulk RNA-seq to pseudotime</h1>
            
      

      <div class="d-none name"><code>assign-bulk-to-pseudotime.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://andrewmccluskey-uog.github.io/blase/" class="external-link">blase</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://statomics.github.io/tradeSeq/index.html" class="external-link">tradeSeq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">slingshot</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/briandconnelly/ami" class="external-link">ami</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGversion</a></span><span class="op">(</span><span class="st">"3.5.0"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in RNGkind("Mersenne-Twister", "Inversion", "Rounding"): non-uniform</span></span>
<span><span class="co">#&gt; 'Rounding' sampler used</span></span>
<span><span class="va">SEED</span> <span class="op">&lt;-</span> <span class="fl">7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N_CORES</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu">ami</span><span class="fu">::</span><span class="fu"><a href="https://briandconnelly.github.io/ami/reference/ci.html" class="external-link">using_ci</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">N_CORES</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In this vignette, we will use BLASE to try to assign bulk RNA-seq
samples to the pseudotime trajectory in a scRNA-seq data. We need to
have the following to use Blase in this case:</p>
<ul>
<li>A Single Cell Experiment with pseudotime.</li>
<li>A list of genes that explain the trajectory well, in this case
generated by tradeSeq’s <code>fitGAM</code> and
<code>associationTest</code>
</li>
</ul>
<div class="section level2">
<h2 id="setting-up-the-single-cell-experiment">Setting up the Single Cell Experiment<a class="anchor" aria-label="anchor" href="#setting-up-the-single-cell-experiment"></a>
</h2>
<p>First, let’s generate a Single Cell Experiment to use the tool with,
based on the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html" class="external-link">tradeSeq
vignette</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">countMatrix</span>, package <span class="op">=</span> <span class="st">"tradeSeq"</span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crv</span>, package <span class="op">=</span> <span class="st">"tradeSeq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">celltype</span>, package <span class="op">=</span> <span class="st">"tradeSeq"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pseudotime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingPseudotime.html" class="external-link">slingPseudotime</a></span><span class="op">(</span><span class="va">crv</span>, na <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">cellWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingPseudotime.html" class="external-link">slingCurveWeights</a></span><span class="op">(</span><span class="va">crv</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tradeSeq/man/fitGAM.html" class="external-link">fitGAM</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">counts</span>, pseudotime <span class="op">=</span> <span class="va">pseudotime</span>, cellWeights <span class="op">=</span> <span class="va">cellWeights</span>,</span>
<span>    nknots <span class="op">=</span> <span class="fl">6</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="va">N_CORES</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">pseudotime</span> <span class="op">&lt;-</span> <span class="va">pseudotime</span><span class="op">[</span>, <span class="st">"curve1"</span><span class="op">]</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="va">celltype</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="va">celltype</span> <span class="op">!=</span> <span class="st">"Erythrocyte"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, text_by <span class="op">=</span> <span class="st">"celltype"</span>, colour_by <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"pseudotime"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, axis_title <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/setupSCEPlot-1.png" alt="Plots showing a UMAP of the single-cell data, coloured by celltype and pseudotime. The UMAP is shaped like a horizontal line, showing the differentiation of a set of immune cell types." width="700"></p>
</div>
<div class="section level2">
<h2 id="finding-the-most-descriptive-genes-with-tradeseq">Finding the most descriptive genes with tradeSeq<a class="anchor" aria-label="anchor" href="#finding-the-most-descriptive-genes-with-tradeseq"></a>
</h2>
<p>Now we’ll find the genes we want to use. We select the top 200 so
that we can do some parameter tuning below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use consecutive for genes that change over time</span></span>
<span><span class="va">assoRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tradeSeq/man/associationTest.html" class="external-link">associationTest</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    lineages <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    global <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    contrastType <span class="op">=</span> <span class="st">"consecutive"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">genelist</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/get_top_n_genes.html">get_top_n_genes</a></span><span class="op">(</span><span class="va">assoRes</span>, n_genes <span class="op">=</span> <span class="fl">200</span>, lineage <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>TODO: Add the custom BLASE genes method here</p>
</div>
<div class="section level2">
<h2 id="parameter-tuning-for-blase">Parameter Tuning for BLASE<a class="anchor" aria-label="anchor" href="#parameter-tuning-for-blase"></a>
</h2>
<p>When using BLASE, it can be a good idea to tune the number of genes
and bins used.</p>
<ul>
<li>Too few genes can lead to poor fitting</li>
<li>Too many genes can lead to slower execution, without any benefit, as
the less informative genes still need to be checked.</li>
<li>Too few bins can oversimplify the trajectory</li>
<li>Too many bins can lead to too few cells being available for reliable
values.</li>
</ul>
<p>We can do this using the following commands, provided by BLASE:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/find_best_params.html">find_best_params</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    <span class="va">genelist</span>,</span>
<span>    split_by <span class="op">=</span> <span class="st">"pseudotime_range"</span>,</span>
<span>    pseudotime_slot <span class="op">=</span> <span class="st">"pseudotime"</span>,</span>
<span>    bins_count_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>    gene_count_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40</span>, <span class="fl">80</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_find_best_params_results.html">plot_find_best_params_results</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/findBestParamsPlot-1.png" alt="Figures showing the minimum and mean convexity, as well as % confident mappings different numbers of genes and bins calculated." width="700"></p>
<p>It looks like 80 genes and 10 bins will give us good specificity, but
let’s double check. In a non trivial dataset, this may take some
repetition. We ignore the 5 bin result because this might not be enough
resolution - this depends on your dataset. In general, more bins will
reduce specificity because each bin will have a more similar cell
composition. In general, aim for about as many bins as you have
clusters.</p>
<div class="section level3">
<h3 id="inspect-bin-choice">Inspect Bin Choice<a class="anchor" aria-label="anchor" href="#inspect-bin-choice"></a>
</h3>
<p>Now we can check if this is a good fit by showing how other bins from
the SC dataset map using these genes. Ideally, each bin is very
specific, with a high “worst specificity”.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blase_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.BlaseData.html">as.BlaseData</a></span><span class="op">(</span><span class="va">sce</span>, pseudotime_slot <span class="op">=</span> <span class="st">"pseudotime"</span>, n_bins <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/genes-getter.html">genes</a></span><span class="op">(</span><span class="va">blase_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genelist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">80</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#| fig.alt: &gt;</span></span>
<span><span class="co">#|   These plots show the correlations of each bin to itself and all other</span></span>
<span><span class="co">#|   bins. This shows how distinct each bin is compared to others, based on</span></span>
<span><span class="co">#|   the number of bins and genes selected.</span></span>
<span><span class="fu"><a href="../reference/evaluate_parameters.html">evaluate_parameters</a></span><span class="op">(</span><span class="va">blase_data</span>, make_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  0.01060  0.03695 30.00000</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="inspect-genes-choice">Inspect Genes Choice<a class="anchor" aria-label="anchor" href="#inspect-genes-choice"></a>
</h3>
<p>We can also see how the genes change over pseudotime, by plotting
expression of the top genes changing over each pseudotime bin.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/evaluate_top_n_genes.html">evaluate_top_n_genes</a></span><span class="op">(</span><span class="va">blase_data</span><span class="op">)</span></span></code></pre></div>
<p>&lt;img
src=“/home/runner/work/blase/blase/docs/articles/assign-bulk-to-pseudotime_files/figure-html/evaluateTopGenes-1.png”
alt=“Plot showing a the”best” genes selected for use with BLASE,
according to how related their expression is to the pseudotime of the
cell.” width=“700” /&gt;</p>
</div>
</div>
<div class="section level2">
<h2 id="mapping-bulk-samples-to-sc-with-blase">Mapping Bulk Samples to SC with BLASE<a class="anchor" aria-label="anchor" href="#mapping-bulk-samples-to-sc-with-blase"></a>
</h2>
<p>If we’re happy, now we can try to map a bulk sample onto the single
cell. We’ll do this by pseudobulking cell types in the
SingleCellExperiment but in reality you should use a real bulk dataset.
See some of our other articles for examples.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulks_df</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">type</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">bulks_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>        <span class="va">bulks_df</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="va">celltype</span> <span class="op">==</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulks_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="va">multipotent_progenitors_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span></span>
<span>    <span class="va">blase_data</span>, <span class="st">"Multipotent progenitors"</span>, <span class="va">bulks_df</span></span>
<span><span class="op">)</span></span>
<span><span class="va">multipotent_progenitors_result</span></span>
<span><span class="co">#&gt; MappingResult for 'Multipotent progenitors': best_bin=1 correlation=0.982747304266292 top_2_distance=0.0731</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.932760097129518 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="va">basophils_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="st">"Basophils"</span>, <span class="va">bulks_df</span><span class="op">)</span></span>
<span><span class="va">basophils_result</span></span>
<span><span class="co">#&gt; MappingResult for 'Basophils': best_bin=6 correlation=0.941795593061416 top_2_distance=0.003</span></span>
<span><span class="co">#&gt;   Confident Result: FALSE (next max upper  0.981516882263948 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="va">neutrophils_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="st">"Neutrophils"</span>, <span class="va">bulks_df</span><span class="op">)</span></span>
<span><span class="va">neutrophils_result</span></span>
<span><span class="co">#&gt; MappingResult for 'Neutrophils': best_bin=10 correlation=0.98019221753399 top_2_distance=0.0584</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.952078251079002 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span></code></pre></div>
<p>We can also generate this result with <code>map_all_best_bins</code>,
which can also allow for parallelisation of the process for large
numbers of bulks:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/map_all_best_bins.html">map_all_best_bins</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="va">bulks_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; MappingResult for 'Multipotent progenitors': best_bin=1 correlation=0.982747304266292 top_2_distance=0.0731</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.938595771129692 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; MappingResult for 'Monocytes': best_bin=9 correlation=0.980496952648851 top_2_distance=0.0072</span></span>
<span><span class="co">#&gt;   Confident Result: FALSE (next max upper  0.987944601456498 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; MappingResult for 'Neutrophils': best_bin=10 correlation=0.98019221753399 top_2_distance=0.0584</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.949560708965291 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; MappingResult for 'Basophils': best_bin=6 correlation=0.941795593061416 top_2_distance=0.003</span></span>
<span><span class="co">#&gt;   Confident Result: FALSE (next max upper  0.979755090550812 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; MappingResult for 'Megakaryocytes': best_bin=2 correlation=0.852156586966713 top_2_distance=0.0444</span></span>
<span><span class="co">#&gt;   Confident Result: FALSE (next max upper  0.907196547720956 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[6]]</span></span>
<span><span class="co">#&gt; MappingResult for 'GMP': best_bin=4 correlation=0.982887951242382 top_2_distance=0.0338</span></span>
<span><span class="co">#&gt;   Confident Result: FALSE (next max upper  0.968805338274442 )</span></span>
<span><span class="co">#&gt;   with history for scores against 10  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span></code></pre></div>
<div class="section level3">
<h3 id="plotting-heatmap-of-mappings">Plotting Heatmap of Mappings<a class="anchor" aria-label="anchor" href="#plotting-heatmap-of-mappings"></a>
</h3>
<p>To see how well we plotted the differences between these cell types,
we can plot a heatmap of all these results. The neutrophils and
Erythrocytes are confidently mapped, but the GMP population doesn’t map
as well.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_mapping_result_heatmap.html">plot_mapping_result_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">multipotent_progenitors_result</span>,</span>
<span>    <span class="va">basophils_result</span>,</span>
<span>    <span class="va">neutrophils_result</span></span>
<span><span class="op">)</span>, annotate_correlation <span class="op">=</span> <span class="cn">TRUE</span>, text_background <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/plotMappingResults-1.png" alt="A heatmap showing the correlations BLASE found for a subset of cell types in this data. Confident mappings are indicated with asterisks over the mapped tile. This information is also available as text from the results of the previous code block." width="700"></p>
</div>
<div class="section level3">
<h3 id="plotting-detailed-correlation-maps">Plotting Detailed Correlation Maps<a class="anchor" aria-label="anchor" href="#plotting-detailed-correlation-maps"></a>
</h3>
<p>To look into the Basophils mapping, we can plot the correlation over
each bin. Because the the lower boundary of the best selection (bin 3)
is higher than the higher bounds of the next-best bin (bin 4), BLASE
makes a confident call.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_mapping_result_corr.html">plot_mapping_result_corr</a></span><span class="op">(</span><span class="va">basophils_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/plotMappingResultCorr-1.png" alt="A plot of the correlation of each bin with the basophils pseudobulk sample, also showing upper and lower bounds as calculated by BLASE. The best mapping is for bin 3, and bin 4 is the next most similar." width="700"></p>
</div>
<div class="section level3">
<h3 id="plotting-summary-plots-of-mappings">Plotting Summary Plots of Mappings<a class="anchor" aria-label="anchor" href="#plotting-summary-plots-of-mappings"></a>
</h3>
<p>Now we can do some more detailed plotting about where the bins mapped
onto the Single Cell data, and the cell type proportions we expect , but
we need to add the pseudotime bins to the metadata of the SCE. We can
see that the true proportions of cell types in the mapped bin do indeed
map to the cell type we expect.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">binned_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assign_pseudotime_bins.html">assign_pseudotime_bins</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    split_by <span class="op">=</span> <span class="st">"pseudotime_range"</span>,</span>
<span>    n_bins <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    pseudotime_slot <span class="op">=</span> <span class="st">"pseudotime"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_mapping_result.html">plot_mapping_result</a></span><span class="op">(</span></span>
<span>    <span class="va">binned_sce</span>,</span>
<span>    <span class="va">multipotent_progenitors_result</span>,</span>
<span>    group_by_slot <span class="op">=</span> <span class="st">"celltype"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/assignPseudotimeBinsPlot-1.png" alt="An overview of mapping for Multipotent progenitors, including UMAPs of all cells and just those in the mapped pseudotime bin coloured by pseudotime bin, and cell-type. Also includes a line plot of correlation of the mapping over every bin, and a bar plot of the different cell-type populations in that bin, showing that the bulk was correctly mapped to a bin consisting mostly of Multipotent progenitor cells." width="700"></p>
<p>To view the bin population chart in full, use:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bin_population.html">plot_bin_population</a></span><span class="op">(</span><span class="va">binned_sce</span>, <span class="fl">1</span>, group_by_slot <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/plotBinPopulation-1.png" alt="A full size bar plot of the cell-type populations of bin 1." width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     Mersenne-Twister </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rounding </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.3.2             ami_0.2.1                  </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.42.1         scater_1.36.0              </span></span>
<span><span class="co">#&gt;  [5] ggplot2_3.5.2               scran_1.36.0               </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.18.0              slingshot_2.16.0           </span></span>
<span><span class="co">#&gt;  [9] TrajectoryUtils_1.16.1      princurve_2.1.6            </span></span>
<span><span class="co">#&gt; [11] tradeSeq_1.22.0             SingleCellExperiment_1.30.1</span></span>
<span><span class="co">#&gt; [13] SummarizedExperiment_1.38.1 Biobase_2.68.0             </span></span>
<span><span class="co">#&gt; [15] GenomicRanges_1.60.0        GenomeInfoDb_1.44.2        </span></span>
<span><span class="co">#&gt; [17] IRanges_2.42.0              S4Vectors_0.46.0           </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.54.0         generics_0.1.4             </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; [23] blase_0.99.0               </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] pbapply_1.7-4           gridExtra_2.3           rlang_1.1.6            </span></span>
<span><span class="co">#&gt;  [4] magrittr_2.0.3          compiler_4.5.1          mgcv_1.9-3             </span></span>
<span><span class="co">#&gt;  [7] systemfonts_1.2.3       vctrs_0.6.5             pkgconfig_2.0.3        </span></span>
<span><span class="co">#&gt; [10] crayon_1.5.3            fastmap_1.2.0           XVector_0.48.0         </span></span>
<span><span class="co">#&gt; [13] labeling_0.4.3          rmarkdown_2.29          UCSC.utils_1.4.0       </span></span>
<span><span class="co">#&gt; [16] ggbeeswarm_0.7.2        ragg_1.5.0              xfun_0.53              </span></span>
<span><span class="co">#&gt; [19] bluster_1.18.0          cachem_1.1.0            beachmat_2.24.0        </span></span>
<span><span class="co">#&gt; [22] jsonlite_2.0.0          DelayedArray_0.34.1     irlba_2.3.5.1          </span></span>
<span><span class="co">#&gt; [25] parallel_4.5.1          cluster_2.1.8.1         R6_2.6.1               </span></span>
<span><span class="co">#&gt; [28] bslib_0.9.0             RColorBrewer_1.1-3      limma_3.64.3           </span></span>
<span><span class="co">#&gt; [31] boot_1.3-31             jquerylib_0.1.4         Rcpp_1.1.0             </span></span>
<span><span class="co">#&gt; [34] knitr_1.50              FNN_1.1.4.1             Matrix_1.7-3           </span></span>
<span><span class="co">#&gt; [37] splines_4.5.1           igraph_2.1.4            tidyselect_1.2.1       </span></span>
<span><span class="co">#&gt; [40] abind_1.4-8             yaml_2.3.10             viridis_0.6.5          </span></span>
<span><span class="co">#&gt; [43] codetools_0.2-20        lattice_0.22-7          tibble_3.3.0           </span></span>
<span><span class="co">#&gt; [46] withr_3.0.2             evaluate_1.0.5          desc_1.4.3             </span></span>
<span><span class="co">#&gt; [49] pillar_1.11.0           scales_1.4.0            glue_1.8.0             </span></span>
<span><span class="co">#&gt; [52] metapod_1.16.0          tools_4.5.1             BiocNeighbors_2.2.0    </span></span>
<span><span class="co">#&gt; [55] ScaledMatrix_1.16.0     locfit_1.5-9.12         fs_1.6.6               </span></span>
<span><span class="co">#&gt; [58] cowplot_1.2.0           grid_4.5.1              edgeR_4.6.3            </span></span>
<span><span class="co">#&gt; [61] nlme_3.1-168            GenomeInfoDbData_1.2.14 beeswarm_0.4.0         </span></span>
<span><span class="co">#&gt; [64] BiocSingular_1.24.0     vipor_0.4.7             cli_3.6.5              </span></span>
<span><span class="co">#&gt; [67] rsvd_1.0.5              textshaping_1.0.3       S4Arrays_1.8.1         </span></span>
<span><span class="co">#&gt; [70] viridisLite_0.4.2       dplyr_1.1.4             uwot_0.2.3             </span></span>
<span><span class="co">#&gt; [73] gtable_0.3.6            sass_0.4.10             digest_0.6.37          </span></span>
<span><span class="co">#&gt; [76] SparseArray_1.8.1       ggrepel_0.9.6           dqrng_0.4.1            </span></span>
<span><span class="co">#&gt; [79] htmlwidgets_1.6.4       farver_2.1.2            htmltools_0.5.8.1      </span></span>
<span><span class="co">#&gt; [82] pkgdown_2.1.3           lifecycle_1.0.4         httr_1.4.7             </span></span>
<span><span class="co">#&gt; [85] statmod_1.5.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrew McCluskey, Toby Kettlewell, Adrian M. Smith, Rhiannon Khundu, David A. Gunn, Thomas D. Otto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
