<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="blase">
<title>Assigning bulk RNA-seq to pseudotime • blase</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assigning bulk RNA-seq to pseudotime">
<meta property="og:description" content="blase">
<meta property="og:image" content="https://andrewmccluskey-uog.github.io/BLASE/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">blase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Vignettes</h6>
    <a class="dropdown-item" href="../articles/How-Blase-Works.html">How BLASE Works</a>
    <a class="dropdown-item" href="../articles/DE_with_blase.html">DE with Blase</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Verifying BLASE Works</h6>
    <a class="dropdown-item" href="../articles/compare_simulated_data.html">Simulated Data Workthrough</a>
    <a class="dropdown-item" href="../articles/assign-bulk-to-pseudotime.html">Assign SC Pseudobulks To SC Pseudotime</a>
    <a class="dropdown-item" href="../articles/BLASE-for-Hematopoesis.html">Weng 2024 Haematopoiesis</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Selecting Genes for Plasmodium sp.</h6>
    <a class="dropdown-item" href="../articles/gene_selection_comparison.html">Comparing Gene Selection Methods In Depth</a>
    <a class="dropdown-item" href="../articles/falciparum_lifecycle.html">P. Falciparum Lifecycle - Comparing Bozdech and TradeSeq</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Data Generation</h6>
    <a class="dropdown-item" href="../articles/data_generation/generate-MCA-pb-object.html">Malaria Cell Atlas PB</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-MCA-pf-object.html">Malaria Cell Atlas PF</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-weng-2024-haematopoiesis-object.html">Weng 2024 Haematopoiesis Test Dataset</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-simulated-data-object.html">Simulation Data</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-zhang-et-al-2021-bulk-FPKM-object.html">Generating Zhang et al. 2021 bulk FPKM Dataframe</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Assigning bulk RNA-seq to pseudotime</h1>
            
      
      
      <div class="d-none name"><code>assign-bulk-to-pseudotime.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://andrewmccluskey-uog.github.io/BLASE/">blase</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://statomics.github.io/tradeSeq/index.html" class="external-link">tradeSeq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">slingshot</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/briandconnelly/ami" class="external-link">ami</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGversion</a></span><span class="op">(</span><span class="st">"3.5.0"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in RNGkind("Mersenne-Twister", "Inversion", "Rounding"): non-uniform</span></span>
<span><span class="co">#&gt; 'Rounding' sampler used</span></span>
<span><span class="va">SEED</span> <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N_CORES</span> <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu">ami</span><span class="fu">::</span><span class="fu"><a href="https://briandconnelly.github.io/ami/reference/ci.html" class="external-link">using_ci</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">N_CORES</span> <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In this vignette, we will use BLASE to try to assign bulk RNA-seq
samples to the pseudotime trajectory in a scRNA-seq data. We need to
have the following to use Blase in this case:</p>
<ul>
<li>A Single Cell Experiment with pseudotime.</li>
<li>A list of genes that explain the trajectory well, in this case
generated by tradeSeq’s <code>fitGAM</code> and
<code>associationTest</code>
</li>
</ul>
<div class="section level2">
<h2 id="setting-up-the-single-cell-experiment">Setting up the Single Cell Experiment<a class="anchor" aria-label="anchor" href="#setting-up-the-single-cell-experiment"></a>
</h2>
<p>First, let’s generate a Single Cell Experiment to use the tool with,
based on the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html" class="external-link">tradeSeq
vignette</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">countMatrix</span>, package <span class="op">=</span> <span class="st">"tradeSeq"</span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crv</span>, package <span class="op">=</span> <span class="st">"tradeSeq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">celltype</span>, package <span class="op">=</span> <span class="st">"tradeSeq"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pseudotime</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingPseudotime.html" class="external-link">slingPseudotime</a></span><span class="op">(</span><span class="va">crv</span>, na <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">cellWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingPseudotime.html" class="external-link">slingCurveWeights</a></span><span class="op">(</span><span class="va">crv</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tradeSeq/man/fitGAM.html" class="external-link">fitGAM</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, pseudotime <span class="op">=</span> <span class="va">pseudotime</span>, cellWeights <span class="op">=</span> <span class="va">cellWeights</span>,</span>
<span>                 nknots <span class="op">=</span> <span class="fl">6</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, BPPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="va">N_CORES</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">pseudotime</span> <span class="op">=</span> <span class="va">pseudotime</span><span class="op">[</span>,<span class="st">"curve1"</span><span class="op">]</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">celltype</span> <span class="op">=</span> <span class="va">celltype</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="va">celltype</span> <span class="op">!=</span> <span class="st">'Erythrocyte'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#sce &lt;- computeSumFactors(sce, clusters=quickCluster(sce))</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, text_by<span class="op">=</span><span class="st">"celltype"</span>, colour_by<span class="op">=</span><span class="st">"celltype"</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"pseudotime"</span><span class="op">)</span>,</span>
<span>      ncol<span class="op">=</span><span class="fl">1</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="finding-the-most-descriptive-genes-with-tradeseq">Finding the most descriptive genes with tradeSeq<a class="anchor" aria-label="anchor" href="#finding-the-most-descriptive-genes-with-tradeseq"></a>
</h2>
<p>Now we’ll find the genes we want to use. We select the top 200 so
that we can do some parameter tuning below.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use consecutive for genes that change over time</span></span>
<span><span class="va">assoRes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/tradeSeq/man/associationTest.html" class="external-link">associationTest</a></span><span class="op">(</span><span class="va">sce</span>, lineages<span class="op">=</span><span class="cn">T</span>, global<span class="op">=</span><span class="cn">F</span>, contrastType<span class="op">=</span><span class="st">"consecutive"</span><span class="op">)</span></span>
<span><span class="va">genelist</span> <span class="op">=</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/get_top_n_genes.html">get_top_n_genes</a></span><span class="op">(</span><span class="va">assoRes</span>, n_genes <span class="op">=</span> <span class="fl">200</span>, lineage <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="parameter-tuning-for-blase">Parameter Tuning for BLASE<a class="anchor" aria-label="anchor" href="#parameter-tuning-for-blase"></a>
</h2>
<p>When using BLASE, it can be a good idea to tune the number of genes
and bins used.</p>
<ul>
<li>Too few genes can lead to poor fitting</li>
<li>Too many genes can lead to slower execution, without any benefit, as
the less informative genes still need to be checked.</li>
<li>Too few bins can oversimplify the trajectory</li>
<li>Too many bins can lead to too few cells being available for reliable
values.</li>
</ul>
<p>We can do this using the following commands, provided by BLASE:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/find_best_params.html">find_best_params</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">genelist</span>, split_by<span class="op">=</span><span class="st">"pseudotime_range"</span>, pseudotime_slot<span class="op">=</span><span class="st">"pseudotime"</span><span class="op">)</span></span>
<span><span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_find_best_params_results.html">plot_find_best_params_results</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>It looks like 80 genes and 10 bins will give us good specificity, but
let’s double check. In a non trivial dataset, this may take some
repetition. We ignore the 5 bin result because this might not be enough
resolution - this depends on your dataset. In general, more bins will
reduce specificity because each bin will have a more similar cell
composition. In general, aim for about as many bins as you have
clusters.</p>
<div class="section level3">
<h3 id="inspect-bin-choice">Inspect Bin Choice<a class="anchor" aria-label="anchor" href="#inspect-bin-choice"></a>
</h3>
<p>Now we can check if this is a good fit by showing how other bins from
the SC dataset map using these genes. Ideally, each bin is very
specific, with a high “worst specificity”.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blase_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.BlaseData.html">as.BlaseData</a></span><span class="op">(</span><span class="va">sce</span>, pseudotime_slot<span class="op">=</span><span class="st">"pseudotime"</span>, n_bins<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">blase_data</span><span class="op">@</span><span class="va">genes</span> <span class="op">=</span> <span class="va">genelist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">80</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/evaluate_parameters.html">evaluate_parameters</a></span><span class="op">(</span><span class="va">blase_data</span>, make_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] 0.03540000 0.08531667</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="inspect-genes-choice">Inspect Genes Choice<a class="anchor" aria-label="anchor" href="#inspect-genes-choice"></a>
</h3>
<p>We can also see how the genes change over pseudotime, by: 1. Using a
fourier transformation to find the phase of the peak of a gene’s
expression (see Bozdech) 2. Plotting expression in cells over
pseudotime, and genes ordered by their phase in pseudotime.</p>
<p>We can see that the genes do not appear to especially differentially
expressed over pseudotime, but this is just a heuristic, so we will move
on to see how they perform.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">waves</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_waves.html">get_waves</a></span><span class="op">(</span><span class="va">sce</span>, pseudotime_slot<span class="op">=</span><span class="st">"pseudotime"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gene_selection_matrix.html">gene_selection_matrix</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">waves</span>, pseudotime_slot<span class="op">=</span><span class="st">"pseudotime"</span>, target_matrix_size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">blase_data</span><span class="op">@</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_selection_matrix.html">gene_selection_matrix</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">waves</span>, pseudotime_slot<span class="op">=</span><span class="st">"pseudotime"</span>, genes<span class="op">=</span><span class="va">blase_data</span><span class="op">@</span><span class="va">genes</span>, target_matrix_size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">blase_data</span><span class="op">@</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<p>You can also inspect how the top genes change over each pseudotime
bin:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/evaluate_top_n_genes.html">evaluate_top_n_genes</a></span><span class="op">(</span><span class="va">blase_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="mapping-bulk-samples-to-sc-with-blase">Mapping Bulk Samples to SC with BLASE<a class="anchor" aria-label="anchor" href="#mapping-bulk-samples-to-sc-with-blase"></a>
</h2>
<p>If we’re happy, now we can try to map a bulk sample onto the single
cell. We’ll do this by pseudobulking cell types in the
SingleCellExperiment but in reality you should use a real bulk dataset.
See some of our other articles for examples.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulks_df</span> <span class="op">=</span> <span class="fu">DataFrame</span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">type</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">bulks_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">bulks_df</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sce</span>, , <span class="va">celltype</span><span class="op">==</span><span class="va">type</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulks_df</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="va">multipotent_progenitors_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="st">"Multipotent progenitors"</span>, <span class="va">bulks_df</span><span class="op">)</span></span>
<span><span class="va">multipotent_progenitors_result</span></span>
<span><span class="co">#&gt; MappingResult for 'Multipotent progenitors': best_bin=1 correlation=0.993670886075949 top_2_distance=0.1767</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.831997187060478 )</span></span>
<span><span class="co">#&gt;   with history for scores against 6  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="va">basophils_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="st">"Basophils"</span>, <span class="va">bulks_df</span><span class="op">)</span></span>
<span><span class="va">basophils_result</span></span>
<span><span class="co">#&gt; MappingResult for 'Basophils': best_bin=4 correlation=0.949554617909048 top_2_distance=0.0038</span></span>
<span><span class="co">#&gt;   Confident Result: FALSE (next max upper  0.95042194092827 )</span></span>
<span><span class="co">#&gt;   with history for scores against 6  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="va">neutrophils_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="st">"Neutrophils"</span>, <span class="va">bulks_df</span><span class="op">)</span></span>
<span><span class="va">neutrophils_result</span></span>
<span><span class="co">#&gt; MappingResult for 'Neutrophils': best_bin=6 correlation=0.980684481950305 top_2_distance=0.1019</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.89943741209564 )</span></span>
<span><span class="co">#&gt;   with history for scores against 6  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span></code></pre></div>
<div class="section level3">
<h3 id="plotting-heatmap-of-mappings">Plotting Heatmap of Mappings<a class="anchor" aria-label="anchor" href="#plotting-heatmap-of-mappings"></a>
</h3>
<p>To see how well we plotted the differences between these cell types,
we can plot a heatmap of all these results. The neutrophils and
Erythrocytes are confidently mapped, but the GMP population doesn’t map
as well.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_mapping_result_heatmap.html">plot_mapping_result_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">multipotent_progenitors_result</span>, </span>
<span>  <span class="va">basophils_result</span>, </span>
<span>  <span class="va">neutrophils_result</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plotting-detailed-correlation-maps">Plotting Detailed Correlation Maps<a class="anchor" aria-label="anchor" href="#plotting-detailed-correlation-maps"></a>
</h3>
<p>To look into the GMP mapping, we can plot the correlation over each
bin. Because the the lower boundary of the best selection (bin 3) is
higher than the higher bounds of the next-best bin (bin 4), BLASE makes
a confident call.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_mapping_result_corr.html">plot_mapping_result_corr</a></span><span class="op">(</span><span class="va">basophils_result</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plotting-summary-plots-of-mappings">Plotting Summary Plots of Mappings<a class="anchor" aria-label="anchor" href="#plotting-summary-plots-of-mappings"></a>
</h3>
<p>Now we can do some more detailed plotting about where the bins mapped
onto the Single Cell data, and the cell type proportions we expect , but
we need to add the pseudotime bins to the metadata of the SCE. We can
see that the true proportions of cell types in the mapped bin do indeed
map to the cell type we expect.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">binned_sce</span> <span class="op">=</span> <span class="fu"><a href="../reference/assign_pseudotime_bins.html">assign_pseudotime_bins</a></span><span class="op">(</span><span class="va">sce</span>, split_by<span class="op">=</span><span class="st">"pseudotime_range"</span>, n_bins<span class="op">=</span><span class="fl">10</span>, pseudotime_slot<span class="op">=</span><span class="st">"pseudotime"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_mapping_result.html">plot_mapping_result</a></span><span class="op">(</span><span class="va">binned_sce</span>, <span class="va">multipotent_progenitors_result</span>, group_by_slot<span class="op">=</span><span class="st">"celltype"</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>To view the bin population chart in full, use:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bin_population.html">plot_bin_population</a></span><span class="op">(</span><span class="va">binned_sce</span>, <span class="fl">1</span>, group_by_slot<span class="op">=</span><span class="st">"celltype"</span><span class="op">)</span></span></code></pre></div>
<p><img src="assign-bulk-to-pseudotime_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>We can see here that both of these bins have high proportions of the
GMP cell type, which is why Blase could not make a convincing call on
either of these being the true “best” mapping. In this case, since Bin 3
has a larger proportion and number of GMP cells, it is likely the best
mapping, and this is reflected in Blase’s correlation score.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     Mersenne-Twister </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rounding </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ami_0.1.0                   BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt;  [3] scater_1.32.0               ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt;  [5] scran_1.32.0                scuttle_1.14.0             </span></span>
<span><span class="co">#&gt;  [7] slingshot_2.12.0            TrajectoryUtils_1.12.0     </span></span>
<span><span class="co">#&gt;  [9] princurve_2.1.6             tradeSeq_1.18.0            </span></span>
<span><span class="co">#&gt; [11] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [13] Biobase_2.64.0              GenomicRanges_1.56.0       </span></span>
<span><span class="co">#&gt; [15] GenomeInfoDb_1.40.0         IRanges_2.38.0             </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.42.0            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt; [19] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt; [21] blase_0.0.0.9000           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] pbapply_1.7-2             gridExtra_2.3            </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.3               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] compiler_4.4.0            mgcv_1.9-1               </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.26.0 reshape2_1.4.4           </span></span>
<span><span class="co">#&gt;   [9] systemfonts_1.0.6         vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [11] stringr_1.5.1             pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [13] crayon_1.5.2              fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [15] backports_1.4.1           XVector_0.44.0           </span></span>
<span><span class="co">#&gt;  [17] labeling_0.4.3            utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [19] rmarkdown_2.26            UCSC.utils_1.0.0         </span></span>
<span><span class="co">#&gt;  [21] ggbeeswarm_0.7.2          ragg_1.3.1               </span></span>
<span><span class="co">#&gt;  [23] purrr_1.0.2               bluster_1.14.0           </span></span>
<span><span class="co">#&gt;  [25] xfun_0.43                 zlibbioc_1.50.0          </span></span>
<span><span class="co">#&gt;  [27] cachem_1.0.8              beachmat_2.20.0          </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.8.8            highr_0.10               </span></span>
<span><span class="co">#&gt;  [31] DelayedArray_0.30.1       cluster_2.1.6            </span></span>
<span><span class="co">#&gt;  [33] irlba_2.3.5.1             parallel_4.4.0           </span></span>
<span><span class="co">#&gt;  [35] R6_2.5.1                  stringi_1.8.4            </span></span>
<span><span class="co">#&gt;  [37] bslib_0.7.0               RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;  [39] limma_3.60.0              jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [41] Rcpp_1.0.12               knitr_1.46               </span></span>
<span><span class="co">#&gt;  [43] FNN_1.1.4                 Matrix_1.7-0             </span></span>
<span><span class="co">#&gt;  [45] splines_4.4.0             igraph_2.0.3             </span></span>
<span><span class="co">#&gt;  [47] tidyselect_1.2.1          abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [49] yaml_2.3.8                viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [51] codetools_0.2-20          plyr_1.8.9               </span></span>
<span><span class="co">#&gt;  [53] lattice_0.22-6            tibble_3.2.1             </span></span>
<span><span class="co">#&gt;  [55] withr_3.0.0               evaluate_0.23            </span></span>
<span><span class="co">#&gt;  [57] desc_1.4.3                pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [59] checkmate_2.3.1           generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [61] metR_0.15.0               sparseMatrixStats_1.16.0 </span></span>
<span><span class="co">#&gt;  [63] munsell_0.5.1             scales_1.3.0             </span></span>
<span><span class="co">#&gt;  [65] glue_1.7.0                metapod_1.12.0           </span></span>
<span><span class="co">#&gt;  [67] tools_4.4.0               BiocNeighbors_1.22.0     </span></span>
<span><span class="co">#&gt;  [69] data.table_1.15.4         ScaledMatrix_1.12.0      </span></span>
<span><span class="co">#&gt;  [71] locfit_1.5-9.9            fs_1.6.4                 </span></span>
<span><span class="co">#&gt;  [73] cowplot_1.1.3             grid_4.4.0               </span></span>
<span><span class="co">#&gt;  [75] edgeR_4.2.0               colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [77] nlme_3.1-164              GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">#&gt;  [79] beeswarm_0.4.0            BiocSingular_1.20.0      </span></span>
<span><span class="co">#&gt;  [81] vipor_0.4.7               cli_3.6.2                </span></span>
<span><span class="co">#&gt;  [83] rsvd_1.0.5                textshaping_0.3.7        </span></span>
<span><span class="co">#&gt;  [85] fansi_1.0.6               S4Arrays_1.4.0           </span></span>
<span><span class="co">#&gt;  [87] viridisLite_0.4.2         dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [89] uwot_0.2.2                gtable_0.3.5             </span></span>
<span><span class="co">#&gt;  [91] sass_0.4.9                digest_0.6.35            </span></span>
<span><span class="co">#&gt;  [93] dqrng_0.3.2               SparseArray_1.4.3        </span></span>
<span><span class="co">#&gt;  [95] ggrepel_0.9.5             farver_2.1.1             </span></span>
<span><span class="co">#&gt;  [97] htmlwidgets_1.6.4         memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [99] htmltools_0.5.8.1         pkgdown_2.0.9            </span></span>
<span><span class="co">#&gt; [101] lifecycle_1.0.4           httr_1.4.7               </span></span>
<span><span class="co">#&gt; [103] statmod_1.5.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrew McCluskey, Toby Kettlewell, Thomas Otto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
