<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="blase">
<title>BLASE for Hematopoesis • blase</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="BLASE for Hematopoesis">
<meta property="og:description" content="blase">
<meta property="og:image" content="https://andrewmccluskey-uog.github.io/BLASE/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">blase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Vignettes</h6>
    <a class="dropdown-item" href="../articles/How-Blase-Works.html">How BLASE Works</a>
    <a class="dropdown-item" href="../articles/DE_with_blase.html">DE with Blase</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Verifying BLASE Works</h6>
    <a class="dropdown-item" href="../articles/compare_simulated_data.html">Simulated Data Workthrough</a>
    <a class="dropdown-item" href="../articles/assign-bulk-to-pseudotime.html">Assign SC Pseudobulks To SC Pseudotime</a>
    <a class="dropdown-item" href="../articles/BLASE-for-Hematopoesis.html">Weng 2024 Haematopoiesis</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Selecting Genes for Plasmodium sp.</h6>
    <a class="dropdown-item" href="../articles/gene_selection_comparison.html">Comparing Gene Selection Methods In Depth</a>
    <a class="dropdown-item" href="../articles/falciparum_lifecycle.html">P. Falciparum Lifecycle - Comparing Bozdech and TradeSeq</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Data Generation</h6>
    <a class="dropdown-item" href="../articles/data_generation/generate-MCA-pb-object.html">Malaria Cell Atlas PB</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-MCA-pf-object.html">Malaria Cell Atlas PF</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-weng-2024-haematopoiesis-object.html">Weng 2024 Haematopoiesis Test Dataset</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-simulated-data-object.html">Simulation Data</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-zhang-et-al-2021-bulk-FPKM-object.html">Generating Zhang et al. 2021 bulk FPKM Dataframe</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>BLASE for Hematopoesis</h1>
            
      
      
      <div class="d-none name"><code>BLASE-for-Hematopoesis.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://andrewmccluskey-uog.github.io/BLASE/">blase</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built with package 'Matrix' 1.6.5 but the current</span></span>
<span><span class="co">#&gt; version is 1.7.0; it is recomended that you reinstall 'SeuratObject' as</span></span>
<span><span class="co">#&gt; the ABI for 'Matrix' may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">slingshot</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: princurve</span></span>
<span><span class="co">#&gt; Loading required package: TrajectoryUtils</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'matrixStats'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     count</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SeuratObject':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:fs':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     path</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,</span></span>
<span><span class="co">#&gt;     tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, rename</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'IRanges'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     collapse, desc, slice</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:sp':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     %over%</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SummarizedExperiment'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Seurat':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Assays</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SeuratObject':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Assays</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://statomics.github.io/tradeSeq/index.html" class="external-link">tradeSeq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'scran'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:TrajectoryUtils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     createClusterMST</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: viridisLite</span></span></code></pre></div>
<p>First configure R:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGversion</a></span><span class="op">(</span><span class="st">"3.5.0"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in RNGkind("Mersenne-Twister", "Inversion", "Rounding"): non-uniform</span></span>
<span><span class="co">#&gt; 'Rounding' sampler used</span></span>
<span><span class="va">SEED</span> <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N_CORES</span> <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu">ami</span><span class="fu">::</span><span class="fu"><a href="https://briandconnelly.github.io/ami/reference/ci.html" class="external-link">using_ci</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">N_CORES</span> <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level2">
<h2 id="load-single-cell">Load Single Cell<a class="anchor" aria-label="anchor" href="#load-single-cell"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, package<span class="op">=</span><span class="st">"blase"</span><span class="op">)</span></span>
<span><span class="va">processed_weng_2024_haematopoiesis</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 1500 6000 </span></span>
<span><span class="co">#&gt; metadata(1): tradeSeq</span></span>
<span><span class="co">#&gt; assays(4): counts logcounts scaledata normcounts</span></span>
<span><span class="co">#&gt; rownames(1500): EBF1 LTBP1 ... MAK PPP2R3A</span></span>
<span><span class="co">#&gt; rowData names(1): tradeSeq</span></span>
<span><span class="co">#&gt; colnames(6000): AAACAGCCACAGGGAC-3 AAACAGCCACTAAATC-3 ...</span></span>
<span><span class="co">#&gt;   TTTGTTGGTAACTACG-3 TTTGTTGGTTAGGCGT-3</span></span>
<span><span class="co">#&gt; colData names(39): orig.ident nCount_RNA ... tradeSeq crv</span></span>
<span><span class="co">#&gt; reducedDimNames(6): PCA UMAP.RNA ... WNN.UMAP UMAP</span></span>
<span><span class="co">#&gt; mainExpName: RNA</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, color_by<span class="op">=</span><span class="st">"STD.CellType"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-Hematopoesis_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, color_by<span class="op">=</span><span class="st">"slingPseudotime_1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-Hematopoesis_files/figure-html/unnamed-chunk-2-2.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, color_by<span class="op">=</span><span class="st">"STD_Cat2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-Hematopoesis_files/figure-html/unnamed-chunk-2-3.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expected_trajectory</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'Refined.HSC'</span>,</span>
<span>    <span class="st">'HSC'</span>, <span class="co"># Haematopoeitic Stem Cell</span></span>
<span>    <span class="st">'MPP'</span>, <span class="co"># multipotent progenitors</span></span>
<span>    <span class="st">'CMP'</span>, <span class="co"># common myeloid progenitor</span></span>
<span>    <span class="st">'GMP'</span>, <span class="co"># granulocyte–monocyte progenitors</span></span>
<span>    <span class="st">'MDP'</span> <span class="co"># Monocyte-dendritic cell progenitor</span></span>
<span>    <span class="co"># 'Mono' # Monocytes - excluded as unclear if part of true haematopoiesis process or mature and BM resident</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">start_cell_type</span> <span class="op">=</span> <span class="va">expected_trajectory</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">end_cell_type</span> <span class="op">=</span> <span class="va">expected_trajectory</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">expected_trajectory</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">associationTestResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tradeSeq/man/associationTest.html" class="external-link">associationTest</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, lineages<span class="op">=</span><span class="cn">T</span>, global<span class="op">=</span><span class="cn">F</span>, contrastType<span class="op">=</span><span class="st">"consecutive"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">associationTestResult</span><span class="op">)</span></span>
<span><span class="co">#&gt;           waldStat_1 df_1 pvalue_1          </span></span>
<span><span class="co">#&gt; EBF1        579.2976    9        0 0.5087777</span></span>
<span><span class="co">#&gt; LTBP1       278.0267    9        0 0.7489076</span></span>
<span><span class="co">#&gt; XACT              NA   NA       NA 0.6707918</span></span>
<span><span class="co">#&gt; LYZ        2087.4766    9        0 0.3033203</span></span>
<span><span class="co">#&gt; LINC02109         NA   NA       NA 0.7437342</span></span>
<span><span class="co">#&gt; MPO        4006.3656    9        0 0.4596214</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">associationTestResult</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1500</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="set-up-blase">Set up BLASE<a class="anchor" aria-label="anchor" href="#set-up-blase"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genelist</span> <span class="op">=</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/get_top_n_genes.html">get_top_n_genes</a></span><span class="op">(</span><span class="va">associationTestResult</span>, n_genes <span class="op">=</span> <span class="fl">1000</span>, lineage <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/find_best_params.html">find_best_params</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, <span class="va">genelist</span>, split_by<span class="op">=</span><span class="st">"pseudotime_range"</span>, bins_count_range<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span>,<span class="fl">9</span>, <span class="fl">10</span><span class="op">)</span>,gene_count_range<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">20</span>,<span class="fl">40</span>, <span class="fl">60</span>, <span class="fl">80</span>, <span class="fl">100</span>, <span class="fl">120</span>, <span class="fl">160</span>, <span class="fl">240</span>, <span class="fl">360</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_find_best_params_results.html">plot_find_best_params_results</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-Hematopoesis_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blase_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.BlaseData.html">as.BlaseData</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, n_bins<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">blase_data</span><span class="op">@</span><span class="va">genes</span> <span class="op">=</span> <span class="va">genelist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">160</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/evaluate_parameters.html">evaluate_parameters</a></span><span class="op">(</span><span class="va">blase_data</span>, make_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-Hematopoesis_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] 0.0322000 0.1417286</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="generate-bulk-datasets">Generate “bulk” datasets<a class="anchor" aria-label="anchor" href="#generate-bulk-datasets"></a>
</h2>
<p>In this article we will pseudobulk the Single Cell data into bulk
datasets to demonstrate that BLASE can map a given cluster back to
itself.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulks_df</span> <span class="op">=</span> <span class="fu">DataFrame</span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">type</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span><span class="op">$</span><span class="va">STD.CellType</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">bulks_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">bulks_df</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span>, , <span class="va">STD.CellType</span><span class="op">==</span><span class="va">type</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulks_df</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">processed_weng_2024_haematopoiesis</span><span class="op">$</span><span class="va">STD.CellType</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-blase">Run BLASE<a class="anchor" aria-label="anchor" href="#run-blase"></a>
</h2>
<p>We will now verify that BLASE can accurately map these “bulk”
datasets back to the SC clusters</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">type</span> <span class="kw">in</span> <span class="va">expected_trajectory</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span></span>
<span>  <span class="va">results</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">results</span>, <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blase_data</span>, <span class="va">type</span>, <span class="va">bulks_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; [1] "Refined.HSC"</span></span>
<span><span class="co">#&gt; [1] "HSC"</span></span>
<span><span class="co">#&gt; [1] "MPP"</span></span>
<span><span class="co">#&gt; [1] "CMP"</span></span>
<span><span class="co">#&gt; [1] "GMP"</span></span>
<span><span class="co">#&gt; [1] "MDP"</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; MappingResult for 'Refined.HSC': best_bin=2 correlation=0.994670690261338 top_2_distance=0.0121</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.985556076409235 )</span></span>
<span><span class="co">#&gt;   with history for scores against 7  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; MappingResult for 'HSC': best_bin=3 correlation=0.985989687097152 top_2_distance=0.0273</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.961933083323567 )</span></span>
<span><span class="co">#&gt;   with history for scores against 7  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; MappingResult for 'MPP': best_bin=4 correlation=0.963380405484589 top_2_distance=0.0468</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.921806515879526 )</span></span>
<span><span class="co">#&gt;   with history for scores against 7  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; MappingResult for 'CMP': best_bin=4 correlation=0.912562990741826 top_2_distance=0.1004</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.824956052970819 )</span></span>
<span><span class="co">#&gt;   with history for scores against 7  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; MappingResult for 'GMP': best_bin=6 correlation=0.932995429508965 top_2_distance=0.0599</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.88220438298371 )</span></span>
<span><span class="co">#&gt;   with history for scores against 7  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[6]]</span></span>
<span><span class="co">#&gt; MappingResult for 'MDP': best_bin=7 correlation=0.945441228172975 top_2_distance=0.1132</span></span>
<span><span class="co">#&gt;   Confident Result: TRUE (next max upper  0.847770420719559 )</span></span>
<span><span class="co">#&gt;   with history for scores against 7  bins</span></span>
<span><span class="co">#&gt;   Bootstrapped with 200 iterations</span></span>
<span><span class="fu"><a href="../reference/plot_mapping_result_heatmap.html">plot_mapping_result_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-Hematopoesis_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     Mersenne-Twister </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rounding </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] viridis_0.6.5               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt;  [3] scater_1.32.0               ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt;  [5] scran_1.32.0                scuttle_1.14.0             </span></span>
<span><span class="co">#&gt;  [7] BiocParallel_1.38.0         tradeSeq_1.18.0            </span></span>
<span><span class="co">#&gt;  [9] slingshot_2.12.0            TrajectoryUtils_1.12.0     </span></span>
<span><span class="co">#&gt; [11] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [13] Biobase_2.64.0              GenomicRanges_1.56.0       </span></span>
<span><span class="co">#&gt; [15] GenomeInfoDb_1.40.0         IRanges_2.38.0             </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.42.0            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt; [19] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt; [21] princurve_2.1.6             dplyr_1.1.4                </span></span>
<span><span class="co">#&gt; [23] Seurat_5.0.3                SeuratObject_5.0.1         </span></span>
<span><span class="co">#&gt; [25] sp_2.1-4                    fs_1.6.4                   </span></span>
<span><span class="co">#&gt; [27] blase_0.0.0.9000           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22          splines_4.4.0            </span></span>
<span><span class="co">#&gt;   [3] later_1.3.2               tibble_3.2.1             </span></span>
<span><span class="co">#&gt;   [5] polyclip_1.10-6           fastDummies_1.7.3        </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4           edgeR_4.2.0              </span></span>
<span><span class="co">#&gt;   [9] globals_0.16.3            lattice_0.22-6           </span></span>
<span><span class="co">#&gt;  [11] MASS_7.3-60.2             backports_1.4.1          </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3            limma_3.60.0             </span></span>
<span><span class="co">#&gt;  [15] plotly_4.10.4             sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.26            jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [19] yaml_2.3.8                metapod_1.12.0           </span></span>
<span><span class="co">#&gt;  [21] httpuv_1.6.15             sctransform_0.4.1        </span></span>
<span><span class="co">#&gt;  [23] spam_2.10-0               spatstat.sparse_3.0-3    </span></span>
<span><span class="co">#&gt;  [25] reticulate_1.36.1         cowplot_1.1.3            </span></span>
<span><span class="co">#&gt;  [27] pbapply_1.7-2             RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;  [29] abind_1.4-5               zlibbioc_1.50.0          </span></span>
<span><span class="co">#&gt;  [31] Rtsne_0.17                purrr_1.0.2              </span></span>
<span><span class="co">#&gt;  [33] GenomeInfoDbData_1.2.12   ggrepel_0.9.5            </span></span>
<span><span class="co">#&gt;  [35] irlba_2.3.5.1             listenv_0.9.1            </span></span>
<span><span class="co">#&gt;  [37] spatstat.utils_3.0-4      goftest_1.2-3            </span></span>
<span><span class="co">#&gt;  [39] RSpectra_0.16-1           dqrng_0.3.2              </span></span>
<span><span class="co">#&gt;  [41] spatstat.random_3.2-3     fitdistrplus_1.1-11      </span></span>
<span><span class="co">#&gt;  [43] parallelly_1.37.1         pkgdown_2.0.9            </span></span>
<span><span class="co">#&gt;  [45] DelayedMatrixStats_1.26.0 leiden_0.4.3.1           </span></span>
<span><span class="co">#&gt;  [47] codetools_0.2-20          DelayedArray_0.30.1      </span></span>
<span><span class="co">#&gt;  [49] tidyselect_1.2.1          farver_2.1.1             </span></span>
<span><span class="co">#&gt;  [51] UCSC.utils_1.0.0          ScaledMatrix_1.12.0      </span></span>
<span><span class="co">#&gt;  [53] spatstat.explore_3.2-7    jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt;  [55] BiocNeighbors_1.22.0      progressr_0.14.0         </span></span>
<span><span class="co">#&gt;  [57] ggridges_0.5.6            survival_3.5-8           </span></span>
<span><span class="co">#&gt;  [59] systemfonts_1.0.6         tools_4.4.0              </span></span>
<span><span class="co">#&gt;  [61] ragg_1.3.1                ica_1.0-3                </span></span>
<span><span class="co">#&gt;  [63] Rcpp_1.0.12               glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [65] gridExtra_2.3             SparseArray_1.4.3        </span></span>
<span><span class="co">#&gt;  [67] mgcv_1.9-1                xfun_0.43                </span></span>
<span><span class="co">#&gt;  [69] withr_3.0.0               fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [71] bluster_1.14.0            fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [73] digest_0.6.35             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [75] R6_2.5.1                  mime_0.12                </span></span>
<span><span class="co">#&gt;  [77] textshaping_0.3.7         colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [79] scattermore_1.2           tensor_1.5               </span></span>
<span><span class="co">#&gt;  [81] spatstat.data_3.0-4       utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [83] tidyr_1.3.1               generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [85] data.table_1.15.4         httr_1.4.7               </span></span>
<span><span class="co">#&gt;  [87] htmlwidgets_1.6.4         S4Arrays_1.4.0           </span></span>
<span><span class="co">#&gt;  [89] uwot_0.2.2                pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [91] gtable_0.3.5              lmtest_0.9-40            </span></span>
<span><span class="co">#&gt;  [93] XVector_0.44.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [95] dotCall64_1.1-1           scales_1.3.0             </span></span>
<span><span class="co">#&gt;  [97] png_0.1-8                 knitr_1.46               </span></span>
<span><span class="co">#&gt;  [99] reshape2_1.4.4            nlme_3.1-164             </span></span>
<span><span class="co">#&gt; [101] checkmate_2.3.1           metR_0.15.0              </span></span>
<span><span class="co">#&gt; [103] cachem_1.0.8              zoo_1.8-12               </span></span>
<span><span class="co">#&gt; [105] stringr_1.5.1             ami_0.1.0                </span></span>
<span><span class="co">#&gt; [107] KernSmooth_2.23-22        parallel_4.4.0           </span></span>
<span><span class="co">#&gt; [109] miniUI_0.1.1.1            vipor_0.4.7              </span></span>
<span><span class="co">#&gt; [111] desc_1.4.3                pillar_1.9.0             </span></span>
<span><span class="co">#&gt; [113] grid_4.4.0                vctrs_0.6.5              </span></span>
<span><span class="co">#&gt; [115] RANN_2.6.1                promises_1.3.0           </span></span>
<span><span class="co">#&gt; [117] BiocSingular_1.20.0       beachmat_2.20.0          </span></span>
<span><span class="co">#&gt; [119] xtable_1.8-4              cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [121] beeswarm_0.4.0            evaluate_0.23            </span></span>
<span><span class="co">#&gt; [123] locfit_1.5-9.9            cli_3.6.2                </span></span>
<span><span class="co">#&gt; [125] compiler_4.4.0            rlang_1.1.3              </span></span>
<span><span class="co">#&gt; [127] crayon_1.5.2              future.apply_1.11.2      </span></span>
<span><span class="co">#&gt; [129] labeling_0.4.3            plyr_1.8.9               </span></span>
<span><span class="co">#&gt; [131] ggbeeswarm_0.7.2          stringi_1.8.4            </span></span>
<span><span class="co">#&gt; [133] deldir_2.0-4              munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [135] lazyeval_0.2.2            spatstat.geom_3.2-9      </span></span>
<span><span class="co">#&gt; [137] Matrix_1.7-0              RcppHNSW_0.6.0           </span></span>
<span><span class="co">#&gt; [139] patchwork_1.2.0           sparseMatrixStats_1.16.0 </span></span>
<span><span class="co">#&gt; [141] future_1.33.2             statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [143] shiny_1.8.1.1             highr_0.10               </span></span>
<span><span class="co">#&gt; [145] ROCR_1.0-11               igraph_2.0.3             </span></span>
<span><span class="co">#&gt; [147] memoise_2.0.1             bslib_0.7.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrew McCluskey, Toby Kettlewell, Thomas Otto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
