<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>BLASE for annotating scRNA-seq • blase</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="BLASE for annotating scRNA-seq">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Vignettes</h6></li>
    <li><a class="dropdown-item" href="../articles/How-Blase-Works.html">How BLASE Works</a></li>
    <li><a class="dropdown-item" href="../articles/assign-bulk-to-pseudotime.html">Assigning Bulk to Pseudotime</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Articles</h6></li>
    <li><a class="dropdown-item" href="../articles/BLASE-for-annotating-scRNA-seq.html">BLASE for annotating scRNA-seq</a></li>
    <li><a class="dropdown-item" href="../articles/BLASE-for-excluding-developmental-genes-from-bulk-RNA-seq.html">BLASE for excluding developmental genes from bulk RNA-seq</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>BLASE for annotating scRNA-seq</h1>
            
      

      <div class="d-none name"><code>BLASE-for-annotating-scRNA-seq.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://andrewmccluskey-uog.github.io/BLASE/">blase</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGversion</a></span><span class="op">(</span><span class="st">"3.5.0"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in RNGkind("Mersenne-Twister", "Inversion", "Rounding"): non-uniform</span></span>
<span><span class="co">#&gt; 'Rounding' sampler used</span></span>
<span><span class="va">SEED</span> <span class="op">&lt;-</span> <span class="fl">7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N_CORES</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu">ami</span><span class="fu">::</span><span class="fu"><a href="https://briandconnelly.github.io/ami/reference/ci.html" class="external-link">using_ci</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">N_CORES</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span>
<span><span class="va">bpparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="va">N_CORES</span><span class="op">)</span></span></code></pre></div>
<p>This article will show how BLASE can be used for annotating scRNA-seq
data using existing bulk or microarray data. We make use of scRNA-seq <a href="https://www.science.org/doi/10.1126/science.adj4088" class="external-link">Dogga et al.
2024</a> and microarray data from <a href="https://www.nature.com/articles/s41467-018-04966-3" class="external-link">Painter et al.
2018</a>. Code for generating the objects used here is available in the
BLASE reproducibility repository.</p>
<div class="section level2">
<h2 id="load-data">Load Data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>First we’ll load in the data we’re using, pre-prepared from the BLASE
reproducibility repository.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">painter_microarray</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Data/painter_2018_pf.rds"</span><span class="op">)</span></span>
<span><span class="va">dogga_sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Data/MCA_PF_SCE.rds"</span><span class="op">)</span></span></code></pre></div>
<p>We can examine the true lifecycle stages, and also the calculated
pseudotime trajectory (Slingshot).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">dogga_sc</span>, colour_by <span class="op">=</span> <span class="st">"STAGE_HR"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/sc_umaps-1.png" alt="UMAP of Dogga et al. single-cell RNA-seq reference coloured by lifecycle stage, showing the cycle from Rings, Trophozoites, Schizonts, to Rings again." width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#| fig.alt: &gt;</span></span>
<span><span class="co">#|   UMAP of Dogga et al. single-cell RNA-seq reference coloured by pseudotime,</span></span>
<span><span class="co">#|   starting with Rings, and ending with Schizonts.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">dogga_sc</span>, colour_by <span class="op">=</span> <span class="st">"slingPseudotime_1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/sc_umaps-2.png" alt="UMAP of Dogga et al. single-cell RNA-seq reference coloured by lifecycle stage, showing the cycle from Rings, Trophozoites, Schizonts, to Rings again." width="700"></p>
</div>
<div class="section level2">
<h2 id="prepare-blase">Prepare BLASE<a class="anchor" aria-label="anchor" href="#prepare-blase"></a>
</h2>
<p>Now we’ll prepare BLASE for use.</p>
<div class="section level3">
<h3 id="create-blase-data-object">Create BLASE data object<a class="anchor" aria-label="anchor" href="#create-blase-data-object"></a>
</h3>
<p>First, we create the object, giving it the number of bins we want to
use, and how to calculate them.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blase_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.BlaseData.html">as.BlaseData</a></span><span class="op">(</span></span>
<span>    <span class="va">dogga_sc</span>,</span>
<span>    pseudotime_slot <span class="op">=</span> <span class="st">"slingPseudotime_1"</span>,</span>
<span>    n_bins <span class="op">=</span> <span class="fl">48</span>,</span>
<span>    split_by <span class="op">=</span> <span class="st">"cells"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add these bins to the sc metadata</span></span>
<span><span class="va">dogga_sc</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/assign_pseudotime_bins.html">assign_pseudotime_bins</a></span><span class="op">(</span><span class="va">dogga_sc</span>,</span>
<span>    pseudotime_slot <span class="op">=</span> <span class="st">"slingPseudotime_1"</span>,</span>
<span>    n_bins <span class="op">=</span> <span class="fl">48</span>,</span>
<span>    split_by <span class="op">=</span> <span class="st">"cells"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-genes">Select Genes<a class="anchor" aria-label="anchor" href="#select-genes"></a>
</h3>
<p>Now we will select the genes we want to use, using BLASE’s gene
peakedness metric.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_peakedness_info</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/calculate_gene_peakedness.html">calculate_gene_peakedness</a></span><span class="op">(</span></span>
<span>    <span class="va">dogga_sc</span>,</span>
<span>    window_pct <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    knots <span class="op">=</span> <span class="fl">18</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="va">bpparam</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes_to_use</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/gene_peakedness_spread_selection.html">gene_peakedness_spread_selection</a></span><span class="op">(</span></span>
<span>    <span class="va">dogga_sc</span>,</span>
<span>    <span class="va">gene_peakedness_info</span>,</span>
<span>    genes_per_bin <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    n_gene_bins <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_peakedness_info</span><span class="op">)</span></span>
<span><span class="co">#&gt;              gene peak_pseudotime mean_in_window mean_out_window     ratio</span></span>
<span><span class="co">#&gt; 72  PF3D7-1401100        9.811180       5.946666        1.576877  3.771165</span></span>
<span><span class="co">#&gt; 57  PF3D7-1401200        7.767184       4.928158        1.981323  2.487307</span></span>
<span><span class="co">#&gt; 59  PF3D7-1401300        8.039717       2.207782        1.285688  1.717199</span></span>
<span><span class="co">#&gt; 721 PF3D7-1401400        9.811180     942.073177       49.944485 18.862406</span></span>
<span><span class="co">#&gt; 15  PF3D7-1401600        2.043996      96.706165        8.207937 11.782031</span></span>
<span><span class="co">#&gt; 19  PF3D7-1402200        2.589061       3.516862        1.549402  2.269819</span></span>
<span><span class="co">#&gt;     window_start window_end deviance_explained</span></span>
<span><span class="co">#&gt; 72      9.470514  10.151846         0.16478147</span></span>
<span><span class="co">#&gt; 57      7.426518   8.107850         0.19846899</span></span>
<span><span class="co">#&gt; 59      7.699051   8.380383         0.04479178</span></span>
<span><span class="co">#&gt; 721     9.470514  10.151846         0.27186813</span></span>
<span><span class="co">#&gt; 15      1.703330   2.384662         0.64855890</span></span>
<span><span class="co">#&gt; 19      2.248395   2.929727         0.17093181</span></span></code></pre></div>
<p>By using the <code>gene_peakedness_spread_selection</code> function,
we can ensure that genes with high ratios are selected from throughout
the trajectory.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gene_peakedness_info</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">peak_pseudotime</span>, y <span class="op">=</span> <span class="va">ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"All genes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/plot_gene_peakedness-1.png" alt="Scatter plot showing gene peakedness ratios for all genes, ordered by pseudotime." width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gene_peakedness_selected_genes</span> <span class="op">&lt;-</span> <span class="va">gene_peakedness_info</span><span class="op">[</span></span>
<span>    <span class="va">gene_peakedness_info</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">genes_to_use</span>,</span>
<span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gene_peakedness_selected_genes</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">peak_pseudotime</span>, y <span class="op">=</span> <span class="va">ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Selected genes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/plot_gene_peakedness_selection-1.png" alt="Scatter plot showing gene peakedness ratios for genes selected by BLASE spread selection, ordered by pseudotime." width="700"></p>
<p>Here, we add them to the BLASE object for mapping with.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genes-getter.html">genes</a></span><span class="op">(</span><span class="va">blase_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes_to_use</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="calculate-mappings">Calculate Mappings<a class="anchor" aria-label="anchor" href="#calculate-mappings"></a>
</h2>
<p>Now we can perform the actual mapping step, and review the
results.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapping_results</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/map_all_best_bins.html">map_all_best_bins</a></span><span class="op">(</span></span>
<span>    <span class="va">blase_data</span>,</span>
<span>    <span class="va">painter_microarray</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="va">bpparam</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_mapping_result_heatmap.html">plot_mapping_result_heatmap</a></span><span class="op">(</span><span class="va">mapping_results</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/mappingPlot-1.png" alt="Heatmap of mapping correlations of the Painter et al. data onto the Dogga et al. scRNA-seq." width="700"></p>
</div>
<div class="section level2">
<h2 id="transfer-mappings">Transfer Mappings<a class="anchor" aria-label="anchor" href="#transfer-mappings"></a>
</h2>
<p>In the Painter et al. paper, the expected lifecycle stages are as
follows: | Lifecycle Stage | Painter et al. HPI | |—————–|——————–| |
Ring | 0-21 | | Trophozoite | 16-32 | | Schizont | 33-48 |</p>
<p>We can use this to transfer back to the scRNA-seq as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">dogga_sc</span>, colour_by <span class="op">=</span> <span class="st">"pseudotime_bin"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/plot_pseudotime_bins-1.png" alt="A UMAP of the reference scRNA-seq data coloured by pseudotime bin, as calculated by BLASE." width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotated_sce</span> <span class="op">&lt;-</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/annotate_sce.html">annotate_sce</a></span><span class="op">(</span><span class="va">dogga_sc</span>, <span class="va">mapping_results</span>, include_stats <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">annotated_sce</span><span class="op">$</span><span class="va">BLASE_Annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">" hpi"</span>, <span class="st">""</span>, <span class="va">annotated_sce</span><span class="op">$</span><span class="va">BLASE_Annotation</span><span class="op">)</span></span>
<span><span class="va">annotated_sce</span><span class="op">$</span><span class="va">BLASE_Annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">annotated_sce</span><span class="op">$</span><span class="va">BLASE_Annotation</span><span class="op">)</span></span>
<span></span>
<span><span class="va">annotation_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">annotated_sce</span>, colour_by <span class="op">=</span> <span class="st">"BLASE_Annotation"</span><span class="op">)</span></span>
<span><span class="va">corr_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">annotated_sce</span>, colour_by <span class="op">=</span> <span class="st">"BLASE_Annotation_Correlation"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Correlation"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation_plot</span> <span class="op">+</span> <span class="va">corr_plot</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/transfer_mappings_to_scPlot-1.png" alt="UMAP of Dogga et al. single-cell RNA-seq reference coloured by best matching bulk RNA-seq sample (hpi). Beside it is a UMAP of the correlations of these mappings." width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotated_sce</span><span class="op">$</span><span class="va">BLASE_Annotation_transfer</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">best_bulk</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">annotated_sce</span><span class="op">$</span><span class="va">BLASE_Annotation</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">mask</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">annotated_sce</span><span class="op">)</span><span class="op">[</span>, <span class="st">"BLASE_Annotation"</span><span class="op">]</span> <span class="op">==</span> <span class="va">best_bulk</span></span>
<span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">best_bulk</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">0</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">annotated_sce</span><span class="op">[</span>, <span class="va">mask</span><span class="op">]</span><span class="op">$</span><span class="va">BLASE_Annotation_transfer</span> <span class="op">&lt;-</span> <span class="st">"Ring"</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">best_bulk</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">16</span><span class="op">:</span><span class="fl">21</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">annotated_sce</span><span class="op">[</span>, <span class="va">mask</span><span class="op">]</span><span class="op">$</span><span class="va">BLASE_Annotation_transfer</span> <span class="op">&lt;-</span> <span class="st">"Ring/Trophozoite"</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">best_bulk</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">22</span><span class="op">:</span><span class="fl">32</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">annotated_sce</span><span class="op">[</span>, <span class="va">mask</span><span class="op">]</span><span class="op">$</span><span class="va">BLASE_Annotation_transfer</span> <span class="op">&lt;-</span> <span class="st">"Trophozoite"</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">best_bulk</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">33</span><span class="op">:</span><span class="fl">48</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">annotated_sce</span><span class="op">[</span>, <span class="va">mask</span><span class="op">]</span><span class="op">$</span><span class="va">BLASE_Annotation_transfer</span> <span class="op">&lt;-</span> <span class="st">"Schizont"</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">annotated_sce</span>, colour_by <span class="op">=</span> <span class="st">"BLASE_Annotation_transfer"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">annotated_sce</span>, colour_by <span class="op">=</span> <span class="st">"STAGE_LR"</span><span class="op">)</span></span></code></pre></div>
<p><img src="BLASE-for-annotating-scRNA-seq_files/figure-html/transfer_painter_stages_to_scPlot-1.png" alt="UMAP of Dogga et al. single-cell RNA-seq reference coloured the original mappings from Dogga et al. and by best matching bulk RNA-seq sample, and then named by the expected cell type ( from annotations in Painter et al.). The mappings are broadly the same." width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     Mersenne-Twister </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rounding </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.3.2             blase_0.99.0               </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.42.1         scater_1.36.0              </span></span>
<span><span class="co">#&gt;  [5] ggplot2_3.5.2               scuttle_1.18.0             </span></span>
<span><span class="co">#&gt;  [7] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.68.0              GenomicRanges_1.60.0       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.44.2         IRanges_2.42.0             </span></span>
<span><span class="co">#&gt; [13] S4Vectors_0.46.0            BiocGenerics_0.54.0        </span></span>
<span><span class="co">#&gt; [15] generics_0.1.4              MatrixGenerics_1.20.0      </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1         viridisLite_0.4.2        ami_0.2.1               </span></span>
<span><span class="co">#&gt;  [4] dplyr_1.1.4              vipor_0.4.7              farver_2.1.2            </span></span>
<span><span class="co">#&gt;  [7] viridis_0.6.5            fastmap_1.2.0            digest_0.6.37           </span></span>
<span><span class="co">#&gt; [10] rsvd_1.0.5               lifecycle_1.0.4          magrittr_2.0.3          </span></span>
<span><span class="co">#&gt; [13] compiler_4.5.1           rlang_1.1.6              sass_0.4.10             </span></span>
<span><span class="co">#&gt; [16] tools_4.5.1              yaml_2.3.10              knitr_1.50              </span></span>
<span><span class="co">#&gt; [19] labeling_0.4.3           S4Arrays_1.8.1           htmlwidgets_1.6.4       </span></span>
<span><span class="co">#&gt; [22] DelayedArray_0.34.1      RColorBrewer_1.1-3       abind_1.4-8             </span></span>
<span><span class="co">#&gt; [25] withr_3.0.2              desc_1.4.3               grid_4.5.1              </span></span>
<span><span class="co">#&gt; [28] beachmat_2.24.0          scales_1.4.0             cli_3.6.5               </span></span>
<span><span class="co">#&gt; [31] rmarkdown_2.29           crayon_1.5.3             ragg_1.4.0              </span></span>
<span><span class="co">#&gt; [34] httr_1.4.7               ggbeeswarm_0.7.2         cachem_1.1.0            </span></span>
<span><span class="co">#&gt; [37] splines_4.5.1            parallel_4.5.1           XVector_0.48.0          </span></span>
<span><span class="co">#&gt; [40] vctrs_0.6.5              boot_1.3-31              Matrix_1.7-3            </span></span>
<span><span class="co">#&gt; [43] jsonlite_2.0.0           BiocSingular_1.24.0      BiocNeighbors_2.2.0     </span></span>
<span><span class="co">#&gt; [46] ggrepel_0.9.6            irlba_2.3.5.1            beeswarm_0.4.0          </span></span>
<span><span class="co">#&gt; [49] systemfonts_1.2.3        jquerylib_0.1.4          glue_1.8.0              </span></span>
<span><span class="co">#&gt; [52] pkgdown_2.1.3            codetools_0.2-20         cowplot_1.2.0           </span></span>
<span><span class="co">#&gt; [55] gtable_0.3.6             UCSC.utils_1.4.0         ScaledMatrix_1.16.0     </span></span>
<span><span class="co">#&gt; [58] tibble_3.3.0             pillar_1.11.0            htmltools_0.5.8.1       </span></span>
<span><span class="co">#&gt; [61] GenomeInfoDbData_1.2.14  R6_2.6.1                 sparseMatrixStats_1.20.0</span></span>
<span><span class="co">#&gt; [64] textshaping_1.0.2        evaluate_1.0.5           lattice_0.22-7          </span></span>
<span><span class="co">#&gt; [67] bslib_0.9.0              Rcpp_1.1.0               gridExtra_2.3           </span></span>
<span><span class="co">#&gt; [70] SparseArray_1.8.1        nlme_3.1-168             mgcv_1.9-3              </span></span>
<span><span class="co">#&gt; [73] xfun_0.53                fs_1.6.6                 pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrew McCluskey, Toby Kettlewell, Adrian M. Smith, Rhiannon Khundu, David A. Gunn, Thomas D. Otto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
