<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="blase">
<title>P. Falciparum Lifecycle - Comparing Bozdech and TradeSeq • blase</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="P. Falciparum Lifecycle - Comparing Bozdech and TradeSeq">
<meta property="og:description" content="blase">
<meta property="og:image" content="https://andrewmccluskey-uog.github.io/BLASE/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">blase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Vignettes</h6>
    <a class="dropdown-item" href="../articles/How-Blase-Works.html">How BLASE Works</a>
    <a class="dropdown-item" href="../articles/DE_with_blase.html">DE with Blase</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Verifying BLASE Works ---</h6>
    <a class="dropdown-item" href="../articles/compare_simulated_data.html">Simulated Data Workthrough</a>
    <a class="dropdown-item" href="../articles/assign-bulk-to-pseudotime.html">Assign SC Pseudobulks To SC Pseudotime</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Selecting Genes for Plasmodium sp.</h6>
    <a class="dropdown-item" href="../articles/gene_selection_comparison.html">Comparing Gene Selection Methods In Depth</a>
    <a class="dropdown-item" href="../articles/falciparum_lifecycle.html">P. Falciparum Lifecycle - Comparing Bozdech and TradeSeq</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Data Generation</h6>
    <a class="dropdown-item" href="../articles/data_generation/generate-MCA-pb-object.html">Malaria Cell Atlas PB</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-MCA-pf-object.html">Malaria Cell Atlas PF</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-weng-2024-haematopoiesis-object.html">Weng 2024 Haematopoiesis Test Dataset</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-simulated-data-object.html">Simulation Data</a>
    <a class="dropdown-item" href="../articles/data_generation/generate-zhang-et-al-2021-bulk-FPKM-object.html">Generating Zhang et al. 2021 bulk FPKM Dataframe</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>P. Falciparum Lifecycle - Comparing Bozdech and TradeSeq</h1>
            
      
      
      <div class="d-none name"><code>falciparum_lifecycle.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">slingshot</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://statomics.github.io/tradeSeq/index.html" class="external-link">tradeSeq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://andrewmccluskey-uog.github.io/BLASE/">blase</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">utils</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/briandconnelly/ami" class="external-link">ami</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGversion</a></span><span class="op">(</span><span class="st">"3.5.0"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in RNGkind("Mersenne-Twister", "Inversion", "Rounding"): non-uniform</span></span>
<span><span class="co">#&gt; 'Rounding' sampler used</span></span>
<span><span class="va">SEED</span> <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N_CORES</span> <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu">ami</span><span class="fu">::</span><span class="fu"><a href="https://briandconnelly.github.io/ami/reference/ci.html" class="external-link">using_ci</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">N_CORES</span> <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In this article we will use BLASE to map <em>Plasmodium
falciparum</em> microarray data of the asexual lifecycle (<a href="https://doi.org/10.1038/s41467-018-04966-3" class="external-link">Painter et
al. 2018</a>, 3d7 strain) onto the Malaria Cell Atlas (<a href="https://www.science.org/doi/10.1126/science.adj4088" class="external-link">Dogga et
al. 2023</a>, 7G8 &amp; NF54 strains) to try and annotate it over a 48
hour period.</p>
<p>We will also investigate the Bozdech method for finding genes that
change over the lifecycle, and compare it with TradeSeq’s results.</p>
<div class="section level2">
<h2 id="load-data">Load Data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">root_dir</span> <span class="op">=</span> <span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/userdir.html" class="external-link">R_user_dir</a></span><span class="op">(</span><span class="st">"BLASE"</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">article_dir</span> <span class="op">=</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">root_dir</span>, <span class="st">"3d7_lifecycle"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">article_dir</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">article_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">bulk_path</span> <span class="op">=</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">article_dir</span>, <span class="st">"3d7_bulk_timepoints"</span>, ext<span class="op">=</span><span class="st">"xlsx"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">bulk_path</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">'https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-018-04966-3/MediaObjects/41467_2018_4966_MOESM4_ESM.xlsx'</span>, <span class="va">bulk_path</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Using cached"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">bulk</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span><span class="va">bulk_path</span>, sheet <span class="op">=</span> <span class="st">"Estimated Total Abundance"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span> <span class="op">=</span> <span class="va">bulk</span><span class="op">$</span><span class="va">`Gene ID`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span>, pattern<span class="op">=</span><span class="st">"_"</span>, replacement<span class="op">=</span><span class="st">"-"</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bulk</span> <span class="op">=</span> <span class="va">bulk</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">bulk</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">bulk</span>, , <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">`Gene ID`</span>, <span class="va">`Annotation`</span>, <span class="va">`Peak Time (hpi)`</span>, <span class="va">`p-value`</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-sc">Prepare SC<a class="anchor" aria-label="anchor" href="#prepare-sc"></a>
</h2>
<p>Here we use the Malaria Cell Atlas’s <em>Plasmodium falciparum</em>
10x data (with some preprocessing described in <a href="generate-MCA-pf-object">generate-MCA-pf-object</a>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, package<span class="op">=</span><span class="st">"blase"</span><span class="op">)</span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"STAGE_LR"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"STAGE_LR"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"DAY"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"DAY"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"STRAIN"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"STRAIN"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"slingPseudotime_1"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, colour_by<span class="op">=</span><span class="st">"slingPseudotime_1"</span><span class="op">)</span>,</span>
<span>  ncol<span class="op">=</span><span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/loadSC-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="get-genes">Get Genes<a class="anchor" aria-label="anchor" href="#get-genes"></a>
</h2>
<div class="section level3">
<h3 id="tradeseq">TradeSeq<a class="anchor" aria-label="anchor" href="#tradeseq"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">associationTestResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tradeSeq/man/associationTest.html" class="external-link">associationTest</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, lineages<span class="op">=</span><span class="cn">T</span>, global<span class="op">=</span><span class="cn">F</span>, contrastType<span class="op">=</span><span class="st">"consecutive"</span><span class="op">)</span></span>
<span><span class="va">genelist</span> <span class="op">=</span> <span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/get_top_n_genes.html">get_top_n_genes</a></span><span class="op">(</span><span class="va">associationTestResult</span>, n_genes <span class="op">=</span> <span class="fl">500</span>, lineage <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_best_params.html">find_best_params</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>,</span>
<span>                       <span class="va">genelist</span>,</span>
<span>                       split_by<span class="op">=</span><span class="st">"pseudotime_range"</span>,</span>
<span>                       pseudotime_slot<span class="op">=</span><span class="st">"slingPseudotime_1"</span>,</span>
<span>                       bins_count_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">14</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                       gene_count_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span><span class="fl">12</span>, <span class="fl">192</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_find_best_params_results.html">plot_find_best_params_results</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res2</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_best_params.html">find_best_params</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>,</span>
<span>                       <span class="va">genelist</span>,</span>
<span>                       split_by<span class="op">=</span><span class="st">"pseudotime_range"</span>,</span>
<span>                       pseudotime_slot<span class="op">=</span><span class="st">"slingPseudotime_1"</span>,</span>
<span>                       bins_count_range <span class="op">=</span> <span class="fl">12</span>,</span>
<span>                       gene_count_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span><span class="fl">12</span>, <span class="fl">192</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_find_best_params_results.html">plot_find_best_params_results</a></span><span class="op">(</span><span class="va">res2</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="get-bozdech-genes">Get Bozdech Genes<a class="anchor" aria-label="anchor" href="#get-bozdech-genes"></a>
</h3>
<p>Bozdech’s paper describes using a fourier transform on the gene
expression of cells ordered by time. This gives the amplitude and phase
of the peak of waves. In this case we assume that many genes important
for lifecycle development are turned on temporarily during the
lifecycle, then turned off, resulting in a clear peak. They then find
the top “power” genes by analysing how strong peaks are versus the
average expression of the gene over pseudotime.</p>
<p>BLASE provides a convenience function to perform this analysis, and
also allows you to select the top genes by amplitude and explained
variance if preferred. See the documentation for
<code><a href="../reference/select_genes_by_fourier_method.html">select_genes_by_fourier_method()</a></code> for more information. It
is noteworthy that BLASE will force these genes to be selected to cover
a wide range of pseudotime by default, instead of simply the top power
genes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">waves</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_waves.html">get_waves</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, <span class="st">"slingPseudotime_1"</span>, n_cores<span class="op">=</span><span class="va">N_CORES</span><span class="op">)</span></span>
<span><span class="va">fourier_power_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/select_genes_by_fourier_method.html">select_genes_by_fourier_method</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, <span class="va">waves</span>, n_groups<span class="op">=</span><span class="fl">100</span>, top_n_per_group<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in .local(x, ...): Fewer genes identified as good matches than</span></span>
<span><span class="co">#&gt; requested. requested=200 found=139</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fourier_power_genes</span><span class="op">)</span></span>
<span><span class="co">#&gt;               amplitude      phase k         r2          gene total_expression</span></span>
<span><span class="co">#&gt; PF3D7-1035900 56.590720 0.06285167 1 0.06903311 PF3D7-1035900        180722.50</span></span>
<span><span class="co">#&gt; PF3D7-0203100 11.329514 0.03413750 1 0.09942970 PF3D7-0203100         42541.57</span></span>
<span><span class="co">#&gt; PF3D7-1035400 64.328026 0.19227314 1 0.06203592 PF3D7-1035400        208335.90</span></span>
<span><span class="co">#&gt; PF3D7-0613900 65.277368 0.19830967 1 0.06787626 PF3D7-0613900        204883.93</span></span>
<span><span class="co">#&gt; PF3D7-0507500  8.015728 0.25552105 1 0.03808634 PF3D7-0507500         30413.86</span></span>
<span><span class="co">#&gt; PF3D7-0210600 20.553304 0.24694346 1 0.06727891 PF3D7-0210600         69061.75</span></span>
<span><span class="co">#&gt;               peak_expression cellcount_in_peak       power</span></span>
<span><span class="co">#&gt; PF3D7-1035900       6061.8055               287 0.009335621</span></span>
<span><span class="co">#&gt; PF3D7-0203100       1322.9882               274 0.008563579</span></span>
<span><span class="co">#&gt; PF3D7-1035400       4213.9056               306 0.015265654</span></span>
<span><span class="co">#&gt; PF3D7-0613900       4585.9112               307 0.014234329</span></span>
<span><span class="co">#&gt; PF3D7-0507500        500.1799               325 0.016025689</span></span>
<span><span class="co">#&gt; PF3D7-0210600       1572.0944               325 0.013073836</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compare-gene-lists">Compare Gene Lists<a class="anchor" aria-label="anchor" href="#compare-gene-lists"></a>
</h3>
<p>Here we generate a heatmap of log counts of the selected genes,
ordered by pseudotime. Genes are ordered by their phase in
pseudotime.</p>
<div class="section level4">
<h4 id="all-genes">All Genes<a class="anchor" aria-label="anchor" href="#all-genes"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_selection_matrix.html">gene_selection_matrix</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, <span class="va">waves</span>, n_cores<span class="op">=</span><span class="va">N_CORES</span>, target_matrix_size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">waves</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="tradeseq-genes">TradeSeq Genes<a class="anchor" aria-label="anchor" href="#tradeseq-genes"></a>
</h4>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_selection_matrix.html">gene_selection_matrix</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, <span class="va">waves</span>, genes<span class="op">=</span><span class="va">genelist</span><span class="op">[</span><span class="fl">0</span><span class="op">:</span><span class="fl">300</span><span class="op">]</span>, n_cores<span class="op">=</span><span class="va">N_CORES</span>, target_matrix_size<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="bozdech-genes">Bozdech Genes<a class="anchor" aria-label="anchor" href="#bozdech-genes"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_selection_matrix.html">gene_selection_matrix</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, <span class="va">waves</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fourier_power_genes</span><span class="op">)</span>, n_cores<span class="op">=</span><span class="va">N_CORES</span>, target_matrix_size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fourier_power_genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="use-each-gene-list">Use each gene list<a class="anchor" aria-label="anchor" href="#use-each-gene-list"></a>
</h2>
<div class="section level3">
<h3 id="using-tradeseq-genes">Using TradeSeq genes<a class="anchor" aria-label="anchor" href="#using-tradeseq-genes"></a>
</h3>
<p>Now we will use the genes generated by TradeSeq.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="va">bins</span> <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="va">processed_MCA_PF_SCE</span> <span class="op">=</span> <span class="fu"><a href="../reference/assign_pseudotime_bins.html">assign_pseudotime_bins</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, pseudotime_slot<span class="op">=</span><span class="st">"slingPseudotime_1"</span>, n_bins<span class="op">=</span><span class="va">bins</span>, split_by<span class="op">=</span><span class="st">"pseudotime_range"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, color<span class="op">=</span><span class="st">"pseudotime_bin"</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, color<span class="op">=</span><span class="st">"STAGE_LR"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, color<span class="op">=</span><span class="st">"pseudotime_bin"</span><span class="op">)</span>,</span>
<span>  ncol<span class="op">=</span><span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-8-2.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blaseData</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.BlaseData.html">as.BlaseData</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, pseudotime_slot<span class="op">=</span><span class="st">"slingPseudotime_1"</span>, n_bins<span class="op">=</span><span class="va">bins</span>, split_by<span class="op">=</span><span class="st">"pseudotime_range"</span><span class="op">)</span></span>
<span><span class="va">blaseData</span><span class="op">@</span><span class="va">genes</span> <span class="op">=</span> <span class="va">genelist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">genes</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/evaluate_parameters.html">evaluate_parameters</a></span><span class="op">(</span><span class="va">blaseData</span>, make_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] 0.1068000 0.1488833</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Map Bulk data onto Single Cell</span></span>
<span><span class="va">mapping_results</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">bulkname</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blaseData</span>, <span class="va">bulkname</span>, <span class="va">bulk</span><span class="op">)</span></span>
<span>  <span class="va">mapping_results</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mapping_results</span>, <span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_mapping_result_heatmap.html">plot_mapping_result_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">mapping_results</span><span class="op">)</span>, annotate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="using-bozdech-genes">Using Bozdech Genes<a class="anchor" aria-label="anchor" href="#using-bozdech-genes"></a>
</h3>
<p>And here we will use the genes generated by Bozdech’s Fourier
method.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">blaseData</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.BlaseData.html">as.BlaseData</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span>, pseudotime_slot<span class="op">=</span><span class="st">"slingPseudotime_1"</span>, n_bins<span class="op">=</span><span class="va">bins</span>, split_by<span class="op">=</span><span class="st">"pseudotime_range"</span><span class="op">)</span></span>
<span><span class="va">blaseData</span><span class="op">@</span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fourier_power_genes</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/evaluate_parameters.html">evaluate_parameters</a></span><span class="op">(</span><span class="va">blaseData</span>, make_plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] 0.095400 0.141875</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Map Bulk data onto Single Cell</span></span>
<span><span class="va">mapping_results_fourier</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">bulkname</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulk</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/map_best_bin.html">map_best_bin</a></span><span class="op">(</span><span class="va">blaseData</span>, <span class="va">bulkname</span>, <span class="va">bulk</span><span class="op">)</span></span>
<span>  <span class="va">mapping_results_fourier</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mapping_results_fourier</span>, <span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">blase</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_mapping_result_heatmap.html">plot_mapping_result_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">mapping_results_fourier</span><span class="op">)</span>, annotate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="falciparum_lifecycle_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="analysing-accuracy">Analysing Accuracy<a class="anchor" aria-label="anchor" href="#analysing-accuracy"></a>
</h2>
<p>The Bulk paper describes the following timepoints HPI matching to
lifecycle stages, in figure 1:</p>
<ul>
<li><p>5-15: Ring</p></li>
<li><p>15-34: Troph</p></li>
<li><p>34-40: Schizont</p></li>
<li><p>40-45: Merozoites</p></li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expected_rings_hpi</span> <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">15</span></span>
<span><span class="va">expected_trophs_hpi</span> <span class="op">=</span> <span class="fl">16</span><span class="op">:</span><span class="fl">34</span></span>
<span><span class="va">expected_schizont_hpi</span> <span class="op">=</span> <span class="fl">35</span><span class="op">:</span><span class="fl">42</span></span></code></pre></div>
<p>Below, we can see how out bins are aligned with lifecycle stages</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">processed_MCA_PF_SCE</span><span class="op">@</span><span class="va">colData</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"STAGE_LR"</span>, <span class="st">"pseudotime_bin"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;              pseudotime_bin</span></span>
<span><span class="co">#&gt; STAGE_LR         1    2    3    4    5    6    7    8    9   10   11   12</span></span>
<span><span class="co">#&gt;   gametocyte     0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;   ring         107   43   35  574  732    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;   schizont     240    0    0    0    0    0    0  258  482  326  273  137</span></span>
<span><span class="co">#&gt;   trophozoite    0    0    0    0  401 1670  512  210    0    0    0    0</span></span></code></pre></div>
<p>Now we can check the accuracy of these techniques, according to prior
knowledge.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_tradeseq</span> <span class="op">=</span> <span class="fu">analyse_mapping_results</span><span class="op">(</span><span class="va">mapping_results</span><span class="op">)</span></span>
<span><span class="va">results_fourier</span> <span class="op">=</span> <span class="fu">analyse_mapping_results</span><span class="op">(</span><span class="va">mapping_results_fourier</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Tradeseq:"</span>, <span class="fu">describe_results</span><span class="op">(</span><span class="va">results_tradeseq</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Tradeseq: 7 wrong out of 38, 1 confident calls wrong out of 13"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Fourier:"</span>, <span class="fu">describe_results</span><span class="op">(</span><span class="va">results_fourier</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Fourier: 1 wrong out of 38, 0 confident calls wrong out of 16"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>We’ve shown here that using Bozdech’s method can give better mappings
than using tradeseq in this case of the <em>P. falciparum</em>
lifecycle. Selecting informative genes is vital for BLASE’s performance,
however this is an non-trivial problem as of now - we reccomend that you
select genes carefully and apply prior knowledge of genes that change
over the trajectory where possible.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     Mersenne-Twister </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rounding </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] fs_1.6.4                    ami_0.1.0                  </span></span>
<span><span class="co">#&gt;  [3] scater_1.32.0               ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt;  [5] blase_0.0.0.9000            BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt;  [7] tradeSeq_1.18.0             slingshot_2.12.0           </span></span>
<span><span class="co">#&gt;  [9] TrajectoryUtils_1.12.0      princurve_2.1.6            </span></span>
<span><span class="co">#&gt; [11] scran_1.32.0                scuttle_1.14.0             </span></span>
<span><span class="co">#&gt; [13] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [15] Biobase_2.64.0              GenomicRanges_1.56.0       </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.40.0         IRanges_2.38.0             </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.42.0            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt; [23] readxl_1.4.3               </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] pbapply_1.7-2             gridExtra_2.3            </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.3               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] compiler_4.4.0            mgcv_1.9-1               </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.26.0 reshape2_1.4.4           </span></span>
<span><span class="co">#&gt;   [9] systemfonts_1.0.6         vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [11] stringr_1.5.1             pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [13] crayon_1.5.2              fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [15] backports_1.4.1           XVector_0.44.0           </span></span>
<span><span class="co">#&gt;  [17] labeling_0.4.3            utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [19] rmarkdown_2.26            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt;  [21] UCSC.utils_1.0.0          ragg_1.3.1               </span></span>
<span><span class="co">#&gt;  [23] purrr_1.0.2               xfun_0.43                </span></span>
<span><span class="co">#&gt;  [25] bluster_1.14.0            zlibbioc_1.50.0          </span></span>
<span><span class="co">#&gt;  [27] cachem_1.0.8              beachmat_2.20.0          </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.8.8            highr_0.10               </span></span>
<span><span class="co">#&gt;  [31] DelayedArray_0.30.1       irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [33] parallel_4.4.0            cluster_2.1.6            </span></span>
<span><span class="co">#&gt;  [35] R6_2.5.1                  stringi_1.8.4            </span></span>
<span><span class="co">#&gt;  [37] bslib_0.7.0               RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;  [39] limma_3.60.0              jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [41] cellranger_1.1.0          Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt;  [43] knitr_1.46                Matrix_1.7-0             </span></span>
<span><span class="co">#&gt;  [45] splines_4.4.0             igraph_2.0.3             </span></span>
<span><span class="co">#&gt;  [47] tidyselect_1.2.1          abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [49] yaml_2.3.8                viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [51] codetools_0.2-20          plyr_1.8.9               </span></span>
<span><span class="co">#&gt;  [53] lattice_0.22-6            tibble_3.2.1             </span></span>
<span><span class="co">#&gt;  [55] withr_3.0.0               evaluate_0.23            </span></span>
<span><span class="co">#&gt;  [57] desc_1.4.3                pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [59] checkmate_2.3.1           generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [61] metR_0.15.0               sparseMatrixStats_1.16.0 </span></span>
<span><span class="co">#&gt;  [63] munsell_0.5.1             scales_1.3.0             </span></span>
<span><span class="co">#&gt;  [65] glue_1.7.0                metapod_1.12.0           </span></span>
<span><span class="co">#&gt;  [67] tools_4.4.0               data.table_1.15.4        </span></span>
<span><span class="co">#&gt;  [69] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0      </span></span>
<span><span class="co">#&gt;  [71] locfit_1.5-9.9            cowplot_1.1.3            </span></span>
<span><span class="co">#&gt;  [73] grid_4.4.0                edgeR_4.2.0              </span></span>
<span><span class="co">#&gt;  [75] colorspace_2.1-0          nlme_3.1-164             </span></span>
<span><span class="co">#&gt;  [77] GenomeInfoDbData_1.2.12   beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [79] BiocSingular_1.20.0       vipor_0.4.7              </span></span>
<span><span class="co">#&gt;  [81] cli_3.6.2                 rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [83] textshaping_0.3.7         fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [85] S4Arrays_1.4.0            viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt;  [87] dplyr_1.1.4               gtable_0.3.5             </span></span>
<span><span class="co">#&gt;  [89] sass_0.4.9                digest_0.6.35            </span></span>
<span><span class="co">#&gt;  [91] ggrepel_0.9.5             SparseArray_1.4.3        </span></span>
<span><span class="co">#&gt;  [93] dqrng_0.3.2               farver_2.1.1             </span></span>
<span><span class="co">#&gt;  [95] htmlwidgets_1.6.4         memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [97] htmltools_0.5.8.1         pkgdown_2.0.9            </span></span>
<span><span class="co">#&gt;  [99] lifecycle_1.0.4           httr_1.4.7               </span></span>
<span><span class="co">#&gt; [101] statmod_1.5.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andrew McCluskey, Toby Kettlewell, Thomas Otto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
