---
title: "Generating Zhang et al. 2021 bulk FPKM Dataframe"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Generating Zhang et al. 2021 bulk FPKM Dataframe}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, eval=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

# Generating Malaria Cell Atlas PF Test Dataset

This article, which is not run for this website, shows how we prepare the [Zhang et al. 2021](https://doi.org/10.1038/s41467-021-24814-1) bulk RNA-seq data set for use with BLASE. It can be used to generate it from scratch, but we provide the processed dataframe with BLASE.

```{r, setup, eval=FALSE}
library(readxl)
library(fs)
library(utils)
library(blase)
```

First configure R:
```{r, randomseed, eval=FALSE}
RNGversion("3.5.0")
SEED <- 7
set.seed(SEED)
```

```{r, concurrency, eval=FALSE}
N_CORES <- 4
if (ami::using_ci()) {
    N_CORES <- 2
}
```

First download the data from the Malaria Cell Atlas
```{r, getData, eval=FALSE}
root_dir <- tools::R_user_dir("BLASE", "data")
article_dir <- path(root_dir, "Generate_zhang_et_al_2021_bulk_fpkm_Object")

if (!dir.exists(article_dir)) {
    dir.create(article_dir, recursive = TRUE)
}

bulk_path <- path(article_dir, "heat_shock_bulks", ext = "xlsx")
if (!file.exists(bulk_path)) {
    download.file("https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-021-24814-1/MediaObjects/41467_2021_24814_MOESM5_ESM.xlsx", bulk_path)
} else {
    print("Using cached")
}

zhang_2021_heat_shock_bulk <- as.data.frame(read_excel(bulk_path, range = "S3A_RNAseq!A4:P2571"))
```

## Clean up the data for use with BLASE
```{r, cleanData, eval=FALSE}
rownames(zhang_2021_heat_shock_bulk) <- zhang_2021_heat_shock_bulk$Transcript_ID

rownames(zhang_2021_heat_shock_bulk) <- gsub(x = rownames(zhang_2021_heat_shock_bulk), pattern = "_", replacement = "-", fixed = TRUE)

zhang_2021_heat_shock_bulk <- subset(zhang_2021_heat_shock_bulk, , -c(`GeneID`, `Transcript_ID`, `Gene.NameSymbol`, `TranscriptProductDescription`))
genes_to_fix_bulk <- rownames(zhang_2021_heat_shock_bulk)[!(rownames(zhang_2021_heat_shock_bulk) %in% sub(x = rownames(zhang_2021_heat_shock_bulk), pattern = "\\.[0-9]", replacement = ""))]
genes_to_fix_bulk <- unique(sub(x = genes_to_fix_bulk, pattern = "\\.[0-9]", replacement = ""))
```

We also remove duplicate transcript ids here.
```{R removeBulkDupTranscripts, eval=FALSE}
new_rows <- data.frame()
rownames_to_remove <- c()

# Do some pre-work to get colSums quickly
bulk_matrix <- data.matrix(zhang_2021_heat_shock_bulk)
bulk_n <- ncol(bulk_matrix)

for (gene in genes_to_fix_bulk) {
    targetRowNames <- rownames(zhang_2021_heat_shock_bulk)[grep(x = rownames(zhang_2021_heat_shock_bulk), pattern = gene)]
    rownames_to_remove <- c(rownames_to_remove, targetRowNames)
    # N = col index, M = row index
    counts <- .colSums(bulk_matrix[targetRowNames, ], m = length(targetRowNames), n = bulk_n)
    counts <- t(data.frame(counts = counts))
    rownames(counts) <- c(gene)
    colnames(counts) <- colnames(zhang_2021_heat_shock_bulk)
    new_rows <- rbind(new_rows, counts)
}

zhang_2021_heat_shock_bulk <- zhang_2021_heat_shock_bulk[!(rownames(zhang_2021_heat_shock_bulk) %in% rownames_to_remove), ]
zhang_2021_heat_shock_bulk <- rbind(zhang_2021_heat_shock_bulk, new_rows)

genes_to_fix_bulk <- rownames(zhang_2021_heat_shock_bulk)[!(rownames(zhang_2021_heat_shock_bulk) %in% sub(x = rownames(zhang_2021_heat_shock_bulk), pattern = "\\.[0-9]", replacement = ""))]
genes_to_fix_bulk <- unique(sub(x = genes_to_fix_bulk, pattern = "\\.[0-9]", replacement = ""))
print(paste("Remaining genes with dups:", genes_to_fix_bulk))

rm(new_rows, rownames_to_remove, bulk_matrix, bulk_n, genes_to_fix_bulk, counts)
gc()
```

## Saving the SCE Object
Now we save the RDS so we can distribute with BLASE.
```{r, saveRDS, eval=FALSE}
usethis::use_data(zhang_2021_heat_shock_bulk, overwrite = TRUE, compress = "bzip2")
```

## Loading the SCE Object
You can load this SCE with:
```{r, loadRDS}
data(zhang_2021_heat_shock_bulk, package = "blase")
zhang_2021_heat_shock_bulk
```

## Session Info
``` {r}
sessionInfo()
```
