---
title: "Generating Malaria Cell Atlas PF Test dataset"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Generating Malaria Cell Atlas PF Test dataset}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, eval=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Generating Malaria Cell Atlas PF Test Dataset

This article, which is not run for this website, shows how we prepare the Malaria Cell Atlas Plasmodium Falciparum ([Dogga et al. 2023](https://www.science.org/doi/10.1126/science.adj4088) 10x single cell data. It can be used to generate it from scratch, but we provide the processed object with BLASE.

```{r, setup, eval=FALSE}
library(readxl)
library(scran)
library(slingshot)
library(tradeSeq)
library(BiocParallel)
library(fs)
library(gridExtra)
library(scater)
library(utils)
library(blase)
library(ami)
```

First configure R:
```{r, randomseed, eval=FALSE}
RNGversion("3.5.0")
SEED = 7
set.seed(SEED)
```

```{r, concurrency, eval=FALSE}
N_CORES = 4
if (ami::using_ci()) {
  N_CORES = 2
}
```

First download the data from the Malaria Cell Atlas
```{r, getData, eval=FALSE}
root_dir = tools::R_user_dir("BLASE", "data")
article_dir = fs::path(root_dir, "Generate_MCA_PF_Object")

if (!dir.exists(article_dir)) {
  dir.create(article_dir, recursive = TRUE)
}

mca_path = path(article_dir, "MCA", ext="zip")
if (!file.exists(mca_path)) {
  download.file('https://www.malariacellatlas.org/downloads/pf.zip', mca_path)
  unzip(mca_path, exdir = paste0(article_dir, '/MCA'))
}else {
  print("Using cached")
}
```

## Prepare SC
Now we read the files and create an SCE.

```{r, loadSC, eval=FALSE}
mca_counts_path = path(article_dir, "MCA", "pf-ch10x-raw", ext="csv")
sc_readcounts = read.csv(mca_counts_path, row.names=1)

mca_annotation_path = path(article_dir, "MCA", "pf-ch10x-data", ext="csv")
sc_annotations = read.csv(mca_annotation_path, row.names=1)

rownames(sc_annotations) = gsub(x = rownames(sc_annotations), pattern="-", replacement=".", fixed=TRUE)
```

Here, we collapse transcripts for the same gene, so that can match some of our bulk datasets better. This is not a reccomended approach in most analyses.
```{r, removeSCDupTranscripts, eval=FALSE}
genes_to_fix = rownames(sc_readcounts)[!(rownames(sc_readcounts) %in% sub(x=rownames(sc_readcounts), pattern="\\.[0-9]", replacement=""))]
genes_to_fix = unique(sub(x=genes_to_fix, pattern="\\.[0-9]", replacement=""))

new_rows = data.frame()
rownames_to_remove = c()

# Do some pre-work to get colSums quickly
sc_readcounts_matrix = data.matrix(sc_readcounts)
sc_readcounts_n = ncol(sc_readcounts_matrix)

for (gene in genes_to_fix) {
  targetRowNames = rownames(sc_readcounts)[grep(x=rownames(sc_readcounts), pattern=gene)]
  rownames_to_remove = c(rownames_to_remove, targetRowNames)
  # N = col index, M = row index
  counts = .colSums(sc_readcounts_matrix[targetRowNames,], m=length(targetRowNames), n=sc_readcounts_n)
  counts = t(data.frame(counts=counts))
  rownames(counts) = c(gene)
  colnames(counts) = colnames(sc_readcounts)
  new_rows = rbind(new_rows, counts)
}

sc_readcounts <- sc_readcounts[ !(rownames(sc_readcounts) %in% rownames_to_remove), ]
sc_readcounts = rbind(sc_readcounts, new_rows)

genes_to_fix = rownames(sc_readcounts)[!(rownames(sc_readcounts) %in% sub(x=rownames(sc_readcounts), pattern="\\.[0-9]", replacement=""))]
genes_to_fix = unique(sub(x=genes_to_fix, pattern="\\.[0-9]", replacement=""))
print(paste("Remaining genes with dups:", genes_to_fix))

rm (new_rows, rownames_to_remove, sc_readcounts_matrix, sc_readcounts_n, genes_to_fix, targetRowNames, gene)
gc()
```

Now we can create a `SingleCellExperiment` object:
```{r, sceCreation, eval=FALSE}
sce = SingleCellExperiment(assays=list(counts=as.matrix(sc_readcounts)), colData=sc_annotations)
rownames(sce) = sub(x=rownames(sce), pattern="\\.[0-9]", replacement="")

sce@colData$STAGE_LR = factor(sce@colData$STAGE_LR)
sce@colData$STAGE_HR = factor(sce@colData$STAGE_HR)
sce@colData$STRAIN = factor(sce@colData$STRAIN)
sce@colData$DAY = factor(sce@colData$DAY)
sce@colData$HOST = factor(sce@colData$HOST)
sce@colData$STAGE_HR2 = factor(sce@colData$STAGE_HR2)
sce@colData$CLUSTER = factor(sce@colData$CLUSTER)
```

We also subset out gametocytes, since we focus on Plasmodium's asexual reproduction in these articles. Then we limit outselves to 10,000 cells, which is plenty for BLASE, and reduces the resource load of this file.
```{r, scSubsetting, eval=FALSE}
sce = subset(sce, , STAGE_LR != 'gametocyte')

# Subsample to reduce memory consumption
cells_to_keep = sample(colnames(sce), size=6000)
sce = sce[,colnames(sce) %in% cells_to_keep]

rm(sc_annotations, sc_readcounts, cells_to_keep)
gc()
```

Now we start annotating and preprocessing the data.
```{r, scPreprocessing, eval=FALSE}
sce <- computeSumFactors(sce)
sce <- logNormCounts(sce)
normcounts(sce) <- exp(logcounts(sce))

dec <- modelGeneVar(sce)
hvg <- getTopHVGs(dec, prop=0.3)
sce = sce[hvg,]

reducedDim(sce, "PCA") = as.matrix(cbind(sce@colData["PC_1"], sce@colData["PC_2"],sce@colData["PC_3"]))
# better separation in the 2nd and 3rd dimension for just asexual stages
reducedDim(sce, "UMAP") = as.matrix(cbind(sce@colData["UMAP_2"], sce@colData["UMAP_3"]))

gridExtra::grid.arrange(
  plotPCA(sce, colour_by="STAGE_HR"),
  plotUMAP(sce, colour_by="STAGE_HR"),
  plotPCA(sce, colour_by="DAY"),
  plotUMAP(sce, colour_by="DAY"),
  plotPCA(sce, colour_by="STRAIN"),
  plotUMAP(sce, colour_by="STRAIN"),
  ncol=2
)
```

## Running Trajectory Analysis

Generate the pseudotime trajectory
```{r, runSlingShot, eval=FALSE}
sce <- slingshot(sce, reducedDim = 'UMAP', clusterLabels="STAGE_HR", start.clus="early ring")
plotUMAP(sce, colour_by="slingPseudotime_1")
```

Evaluate the number of knots required by tradeseq for differential gene expression over pseudotime.
```{r, evaluateKForTradeSeq, eval=FALSE}
evaluateK(sce, k=6:8, parallel=T, BPPARAM=MulticoreParam(N_CORES), nGenes=length(rownames(sce)))
```

Find differentially expressed genes over pseudotime:
```{r, tradeSeq, eval=FALSE}
sce <- fitGAM(sce, parallel=T, BPPARAM=MulticoreParam(N_CORES), nknots=7)
```

The differentially expressed genes can be found using the association test like this in other places:
```{r, associationTest, eval=FALSE}
associationTestResult <- associationTest(sce, lineages=T, global=F, contrastType="consecutive")
head(associationTestResult)
nrow(associationTestResult)
```

## Saving the SCE Object
Now we save the RDS so we can distribute with BLASE.
```{r, saveRDS, eval=FALSE}
processed_MCA_PF_SCE = sce
usethis::use_data(processed_MCA_PF_SCE, overwrite = TRUE, compress = "bzip2")
```

## Loading the SCE Object
You can load this SCE with:
```{r, loadRDS}
data(processed_MCA_PF_SCE, package="blase")
processed_MCA_PF_SCE
```

## Session Info
``` {r}
sessionInfo()
```
